# allos


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

Single-cell RNA sequencing (scRNA-seq) has revolutionized our
understanding of cellular diversity by allowing the study of gene
expression at the individual cell level. However, traditional methods of
quantification obscure the most fine-grained transcriptional layerâ€”the
individual transcripts at sequence-level resolutionâ€”in favor of
gene-level binning. This strategy is useful when the sequencing read
length is short, as it avoids the difficult task of transcript assembly,
which is made even more challenging by the nature of single-cell data.

Traditional scRNA-seq methods often rely on short-read sequencing
technologies, which can limit the ability to resolve full-length
transcripts and isoforms. Long-read sequencing technologies, such as
those provided by Oxford Nanopore and PacBio, overcome this limitation
by producing reads that span entire transcripts. This capability is
crucial for accurately identifying and characterizing novel isoforms and
understanding the complexity of transcriptomes in single cells. By
integrating long-read sequencing with single-cell approaches,
researchers can gain deeper insights into cellular heterogeneity and the
functional implications of transcriptomic variations.

Allos is designed to give users familiar with single-cell analysis, and
indeed object-oriented programming in general, an extended toolkit
wrapping around Scanpy and the wider scverse to facilitate analysis
workflows of this type of data. Allos is the culmination of many smaller
individualized scripts packaged together into a common framework
centered on anndata objects. It also contains many recreations (of
varying faithfulness to their originals\*) of functions proposed by
others in other libraries, languages, or contextsâ€”we by no means intend
to plagiarize these methods but only expand them to a wider audience.

Allos intends to offer a full suite of modules for every step of
single-cell isoform resolution data, from preprocessing, working with
annotations, plotting, identifying differential features, isoform
switches, and more. We hope to expand Allos with the input of our user
base as the field of long-read single-cell further matures.

## Install

``` sh
pip install allos
```

## Basic workflow

The first thing we need to work with Allos is data. Our goal is for
Allos to handle many, if not all, types of single-cell isoform
resolution data, making it platform and protocol agnostic. While it is
important to consider the biases, strengths, limitations, and caveats of
each approach, all methods utilize a transcript matrix from which a gene
matrix can be derived. Allos also allows users to supply their own
custom annotations in the form of a GTF or use a reference annotation.
If a reference annotation is provided, all plotting functions will use
the underlying GTF to retrieve the transcript information.

To help you follow the basic workflow, we provide an easy way to
download one of our test datasets. The Sicelore dataset comprises 1,121
single-cell transcriptomes from two technical replicates derived from an
embryonic day 18 (E18) mouse brain, consisting of 951 cells and 190
cells. The dataset was generated using the 10x Genomics Chromium system
and sequenced with both Oxford Nanopore and Illumina platforms,
producing 322 million Nanopore reads and 70 million Illumina reads for
the 951-cell replicate, and 32 million Nanopore reads with 43 million
Illumina reads for the 190-cell replicate. The dataset captured a median
of 2,427 genes and 6,047 UMIs per cell, enabling the identification of
33,002 annotated transcript isoforms and 4,388 novel isoforms. Major
cell types in the dataset include radial glia, cycling radial glia,
intermediate progenitors, Cajal-Retzius cells, and maturing GABAergic
and glutamatergic neurons, providing a detailed resource for studying
transcriptome-wide alternative splicing, isoform expression, and
sequence diversity during mouse brain development.

``` python
import allos.preprocessing as pp
sicelore_mouse_data = pp.process_mouse_data()
```


    ðŸ”Ž Looking for file at: /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/e18.mouse.clusters.csv
    âœ… File found at: /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/e18.mouse.clusters.csv
    âœ… File already exists at: /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_1.txt.gz

    ðŸ”„ Decompressing /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_1.txt.gz to /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_1.txt...
    âœ… Decompression complete.
    Test data (mouse_1) downloaded successfully
    âœ… File already exists at: /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_2.txt.gz

    ðŸ”„ Decompressing /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_2.txt.gz to /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_2.txt...
    âœ… Decompression complete.
    Test data (mouse_2) downloaded successfully

    /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
      utils.warn_names_duplicates("obs")

In this dataset, we observe the mouse data organized in a
cell-by-transcript matrix format. This structure allows us to examine
the expression levels of individual transcripts across different cells.
Additionally, each transcript is associated with a specific gene,
providing a hierarchical view of gene expression. This organization is
crucial for understanding the relationship between transcripts and their
corresponding genes.

``` python
sicelore_mouse_data
```

    View of AnnData object with n_obs Ã— n_vars = 1109 Ã— 31986
        obs: 'batch', 'cell_type'
        var: 'geneId'

``` python
sicelore_mouse_data.var
```

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }
&#10;    .dataframe tbody tr th {
        vertical-align: top;
    }
&#10;    .dataframe thead th {
        text-align: right;
    }
</style>

|                       | geneId |
|-----------------------|--------|
| transcriptId          |        |
| ENSMUST00000156717.1  | Klc2   |
| ENSMUST00000212520.1  | Capn15 |
| ENSMUST00000025798.12 | Klc2   |
| ENSMUST00000231280.1  | Eva1c  |
| ENSMUST00000039286.4  | Atg5   |
| ...                   | ...    |
| ENSMUST00000127375.1  | Snx11  |
| ENSMUST00000125687.1  | Camta2 |
| ENSMUST00000139277.7  | Atp5o  |
| ENSMUST00000044234.13 | Tgfbr1 |
| ENSMUST00000198403.1  | Ptbp2  |

<p>31986 rows Ã— 1 columns</p>
</div>
