<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="description" content="This module provides tools to identify potential isoform switches within single-cell isoform-resolution datasets. It includes methods based on Dirichlet distributions and Wilcoxon tests for detecting differential isoform usage. These approaches are still under development and intended for exploratory use, with manual confirmation of detected isoform switches being essential.">

<title>Detecting Switches â€“ allos</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-9bc060a4898e0fcb4a0a3e2203846fdc.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="Detecting Switches â€“ allos">
<meta property="og:description" content="This module provides tools to identify potential isoform switches within single-cell isoform-resolution datasets. It includes methods based on Dirichlet distributions and Wilcoxon tests for detecting differential isoform usage. These approaches are still under development and intended for exploratory use, with manual confirmation of detected isoform switches being essential.">
<meta property="og:image" content="https://cobioda.github.io/allos/005_switch_search_files/figure-html/cell-15-output-2.png">
<meta property="og:site_name" content="allos">
<meta property="og:image:height" content="860">
<meta property="og:image:width" content="1517">
<meta name="twitter:title" content="Detecting Switches â€“ allos">
<meta name="twitter:description" content="This module provides tools to identify potential isoform switches within single-cell isoform-resolution datasets. It includes methods based on Dirichlet distributions and Wilcoxon tests for detecting differential isoform usage. These approaches are still under development and intended for exploratory use, with manual confirmation of detected isoform switches being essential.">
<meta name="twitter:image" content="https://cobioda.github.io/allos/005_switch_search_files/figure-html/cell-15-output-2.png">
<meta name="twitter:image-height" content="860">
<meta name="twitter:image-width" content="1517">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">allos</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./switch_search.html">Detecting Switches</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Allos</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./palettes.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Palettes</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./transcript_plots.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Transcript plots</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./transcript_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Transcript Data</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./readers_tests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Readers</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./anndata_iso.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">IsoAdata</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./switch_search.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Detecting Switches</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./gene_report.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Gene Report Module</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./preprocesing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preprocessing</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./visuals.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Visualizations</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./metacells.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Metacells</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#switchsearch" id="toc-switchsearch" class="nav-link active" data-scroll-target="#switchsearch">SwitchSearch</a></li>
  <li><a href="#process_pair_wilcoxon" id="toc-process_pair_wilcoxon" class="nav-link" data-scroll-target="#process_pair_wilcoxon">process_pair_wilcoxon</a></li>
  <li><a href="#process_gene_batch" id="toc-process_gene_batch" class="nav-link" data-scroll-target="#process_gene_batch">process_gene_batch</a></li>
  <li><a href="#compute_batch_size" id="toc-compute_batch_size" class="nav-link" data-scroll-target="#compute_batch_size">compute_batch_size</a></li>
  <li><a href="#chunk_list" id="toc-chunk_list" class="nav-link" data-scroll-target="#chunk_list">chunk_list</a></li>
  <li><a href="#create_iso_adata" id="toc-create_iso_adata" class="nav-link" data-scroll-target="#create_iso_adata">create_iso_adata</a></li>
  <li><a href="#get_top_n_isoforms" id="toc-get_top_n_isoforms" class="nav-link" data-scroll-target="#get_top_n_isoforms">get_top_n_isoforms</a></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/cobioda/allos/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Detecting Switches</h1>
</div>

<div>
  <div class="description">
    This module provides tools to identify potential isoform switches within single-cell isoform-resolution datasets. It includes methods based on Dirichlet distributions and Wilcoxon tests for detecting differential isoform usage. These approaches are still under development and intended for exploratory use, with manual confirmation of detected isoform switches being essential.
  </div>
</div>


<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<pre><code>/data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/fastcore/docscrape.py:230: UserWarning: Unknown section See Also
  else: warn(msg)</code></pre>
<hr>
<p><a href="https://github.com/cobioda/allos/blob/main/allos/switch_search.py#L140" target="_blank" style="float:right; font-size:smaller">source</a></p>
<section id="switchsearch" class="level3">
<h3 class="anchored" data-anchor-id="switchsearch">SwitchSearch</h3>
<blockquote class="blockquote">
<pre><code> SwitchSearch (anndata:anndata._core.anndata.AnnData)</code></pre>
</blockquote>
<p>*An annotated data matrix.</p>
<p>.. figure:: ../_static/img/anndata_schema.svg :width: 260px :align: right :class: dark-light</p>
<p>:class:<code>~anndata.AnnData</code> stores a data matrix :attr:<code>X</code> together with annotations of observations :attr:<code>obs</code> (:attr:<code>obsm</code>, :attr:<code>obsp</code>), variables :attr:<code>var</code> (:attr:<code>varm</code>, :attr:<code>varp</code>), and unstructured annotations :attr:<code>uns</code>.</p>
<p>An :class:<code>~anndata.AnnData</code> object <code>adata</code> can be sliced like a :class:<code>~pandas.DataFrame</code>, for instance <code>adata_subset = adata[:, list_of_variable_names]</code>. :class:<code>~anndata.AnnData</code>â€™s basic structure is similar to Râ€™s ExpressionSet [Huber15]_. If setting an <code>.h5ad</code>-formatted HDF5 backing file <code>.filename</code>, data remains on the disk but is automatically loaded into memory if needed.*</p>
<hr>
<p><a href="https://github.com/cobioda/allos/blob/main/allos/switch_search.py#L56" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="process_pair_wilcoxon" class="level3">
<h3 class="anchored" data-anchor-id="process_pair_wilcoxon">process_pair_wilcoxon</h3>
<blockquote class="blockquote">
<pre><code> process_pair_wilcoxon (adata_norm, cell_labels_column, group1, group2)</code></pre>
</blockquote>
<p><em>Process one pairwise Scanpy Wilcoxon test. Operates on a copy of adata_norm.</em></p>
<hr>
<p><a href="https://github.com/cobioda/allos/blob/main/allos/switch_search.py#L47" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="process_gene_batch" class="level3">
<h3 class="anchored" data-anchor-id="process_gene_batch">process_gene_batch</h3>
<blockquote class="blockquote">
<pre><code> process_gene_batch (filtered_ss, group1, group2, cell_group_column,
                     gene_batch)</code></pre>
</blockquote>
<p><em>Process a batch of gene IDs for the Dirichlet test.</em></p>
<hr>
<p><a href="https://github.com/cobioda/allos/blob/main/allos/switch_search.py#L43" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="compute_batch_size" class="level3">
<h3 class="anchored" data-anchor-id="compute_batch_size">compute_batch_size</h3>
<blockquote class="blockquote">
<pre><code> compute_batch_size (num_items, n_jobs)</code></pre>
</blockquote>
<p><em>Compute a batch size that divides num_items roughly equally among n_jobs workers.</em></p>
<hr>
<p><a href="https://github.com/cobioda/allos/blob/main/allos/switch_search.py#L38" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="chunk_list" class="level3">
<h3 class="anchored" data-anchor-id="chunk_list">chunk_list</h3>
<blockquote class="blockquote">
<pre><code> chunk_list (lst, chunk_size)</code></pre>
</blockquote>
<p><em>Yield successive chunks (batches) of chunk_size from lst.</em></p>
<hr>
<p><a href="https://github.com/cobioda/allos/blob/main/allos/switch_search.py#L660" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="create_iso_adata" class="level3">
<h3 class="anchored" data-anchor-id="create_iso_adata">create_iso_adata</h3>
<blockquote class="blockquote">
<pre><code> create_iso_adata (path:str)</code></pre>
</blockquote>
<p>*Creates an AnnData object containing isoform counts from a Sicelore output file.</p>
<p>Parameters: - path (str): Path to the input file, which is expected to be tab-separated and contain: - <code>transcriptId</code>: Transcript identifiers. - <code>geneId</code>: Gene identifiers. - Transcript counts for individual barcodes (cells) as additional columns.</p>
<p>Returns: - adata_iso (AnnData): An AnnData object with isoform counts.*</p>
<table class="caption-top table">
<colgroup>
<col style="width: 9%">
<col style="width: 38%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th><strong>Type</strong></th>
<th><strong>Details</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>path</td>
<td>str</td>
<td>Path to a tab-separated file, typically the Sicelore output with transcript counts.</td>
</tr>
</tbody>
</table>
<div id="3719063a" class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> allos.readers_tests <span class="im">import</span> <span class="op">*</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> allos.preprocessing <span class="im">as</span> pp</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>combined_mouse_data <span class="op">=</span> process_mouse_data()</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>combined_mouse_data <span class="op">=</span> pp.filter_transcripts_by_abundance(combined_mouse_data,  threshold_pct<span class="op">=</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
ðŸ”Ž Looking for file at: /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/e18.mouse.clusters.csv
âœ… File found at: /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/e18.mouse.clusters.csv
âœ… File already exists at: /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_1.txt.gz

ðŸ”„ Decompressing /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_1.txt.gz to /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_1.txt...
âœ… Decompression complete.
Test data (mouse_1) downloaded successfully
âœ… File already exists at: /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_2.txt.gz

ðŸ”„ Decompressing /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_2.txt.gz to /data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/allos/resources/data/mouse_2.txt...
âœ… Decompression complete.
Test data (mouse_2) downloaded successfully</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>/data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
  utils.warn_names_duplicates("obs")</code></pre>
</div>
</div>
<div id="d7beaf31" class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>combined_mouse_data.obs[<span class="st">'cell_type'</span>].unique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>array(['mature Glutamatergic', 'Cajal-Retzius', 'imature Glutamatergic',
       'intermediate progenitor', 'radial glia', 'cycling radial glia',
       'imature GABAergic', 'mature GABAergic'], dtype=object)</code></pre>
</div>
</div>
<div id="2864493e" class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>ss <span class="op">=</span> SwitchSearch(combined_mouse_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="07710db7" class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the Dirichletâ€“based switch test (will error if data appear logâ€“transformed):</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>dirichlet_results <span class="op">=</span> ss.find_switching_isoforms_dirichlet(cell_group_column<span class="op">=</span><span class="st">'cell_type'</span>,</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                                                         min_count<span class="op">=</span><span class="dv">30</span>, min_log_fold_change<span class="op">=</span><span class="dv">1</span>, n_jobs <span class="op">=</span> <span class="dv">10</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the Wilcoxonâ€“based switch test (will CPMâ€“normalize and logâ€“transform if needed):</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>wilcoxon_results <span class="op">=</span> ss.find_switching_isoforms_wilcoxon(cell_group_column<span class="op">=</span><span class="st">'cell_type'</span>,</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>                                                       min_fdr<span class="op">=</span><span class="fl">0.05</span>, min_log_fold_change<span class="op">=</span><span class="dv">1</span>, n_jobs <span class="op">=</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/tmp/ipykernel_847531/2950237862.py:389: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  marker_df_filtered</code></pre>
</div>
</div>
<div id="1e113cfd" class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>wilcoxon_results.sort_values(by<span class="op">=</span><span class="st">'log_fold_change'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">transcript_id</th>
<th data-quarto-table-cell-role="th">scores</th>
<th data-quarto-table-cell-role="th">log_fold_change</th>
<th data-quarto-table-cell-role="th">pvals</th>
<th data-quarto-table-cell-role="th">pvals_adj</th>
<th data-quarto-table-cell-role="th">group_1</th>
<th data-quarto-table-cell-role="th">group_2</th>
<th data-quarto-table-cell-role="th">contrast</th>
<th data-quarto-table-cell-role="th">gene_id</th>
<th data-quarto-table-cell-role="th">n_cells_group_1</th>
<th data-quarto-table-cell-role="th">n_cells_group_2</th>
<th data-quarto-table-cell-role="th">total_cells</th>
<th data-quarto-table-cell-role="th">p_value</th>
<th data-quarto-table-cell-role="th">direction</th>
<th data-quarto-table-cell-role="th">percent_expressed_group_1</th>
<th data-quarto-table-cell-role="th">percent_expressed_group_2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">1104</td>
<td>ENSMUST00000051477.12</td>
<td>3.983871</td>
<td>1.313683</td>
<td>6.780165e-05</td>
<td>1.978236e-03</td>
<td>cycling radial glia</td>
<td>mature Glutamatergic</td>
<td>cycling radial glia__mature Glutamatergic</td>
<td>Cdc42</td>
<td>117</td>
<td>275</td>
<td>392</td>
<td>1.491337e-02</td>
<td>1.313683</td>
<td>85.470085</td>
<td>72.727273</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5019</td>
<td>ENSMUST00000164181.1</td>
<td>4.931299</td>
<td>1.337557</td>
<td>8.168465e-07</td>
<td>4.777785e-05</td>
<td>mature GABAergic</td>
<td>radial glia</td>
<td>mature GABAergic__radial glia</td>
<td>Myl6</td>
<td>172</td>
<td>68</td>
<td>240</td>
<td>5.727071e-04</td>
<td>1.337557</td>
<td>81.395349</td>
<td>75.000000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6031</td>
<td>ENSMUST00000050970.3</td>
<td>3.715006</td>
<td>1.635736</td>
<td>2.031990e-04</td>
<td>5.826363e-03</td>
<td>mature Glutamatergic</td>
<td>radial glia</td>
<td>mature Glutamatergic__radial glia</td>
<td>Polr1d</td>
<td>275</td>
<td>68</td>
<td>343</td>
<td>3.763291e-02</td>
<td>1.635736</td>
<td>77.454545</td>
<td>61.764706</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">6032</td>
<td>ENSMUST00000110557.4</td>
<td>3.652179</td>
<td>1.638091</td>
<td>2.600250e-04</td>
<td>7.239154e-03</td>
<td>radial glia</td>
<td>mature Glutamatergic</td>
<td>mature Glutamatergic__radial glia</td>
<td>Polr1d</td>
<td>68</td>
<td>275</td>
<td>343</td>
<td>4.529321e-02</td>
<td>-1.638091</td>
<td>79.411765</td>
<td>61.818182</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1478</td>
<td>ENSMUST00000170241.7</td>
<td>4.264503</td>
<td>1.639061</td>
<td>2.003471e-05</td>
<td>3.617426e-03</td>
<td>imature GABAergic</td>
<td>imature Glutamatergic</td>
<td>imature GABAergic__imature Glutamatergic</td>
<td>Clta</td>
<td>141</td>
<td>234</td>
<td>375</td>
<td>2.520287e-02</td>
<td>1.639061</td>
<td>60.283688</td>
<td>45.299145</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1495</td>
<td>ENSMUST00000107849.9</td>
<td>9.634603</td>
<td>8.724378</td>
<td>5.711261e-22</td>
<td>2.846150e-19</td>
<td>mature GABAergic</td>
<td>radial glia</td>
<td>mature GABAergic__radial glia</td>
<td>Clta</td>
<td>172</td>
<td>68</td>
<td>240</td>
<td>2.694471e-17</td>
<td>8.724378</td>
<td>83.139535</td>
<td>7.352941</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1497</td>
<td>ENSMUST00000107849.9</td>
<td>10.725714</td>
<td>8.955872</td>
<td>7.708972e-27</td>
<td>4.268543e-24</td>
<td>mature Glutamatergic</td>
<td>radial glia</td>
<td>mature Glutamatergic__radial glia</td>
<td>Clta</td>
<td>275</td>
<td>68</td>
<td>343</td>
<td>7.447796e-22</td>
<td>8.955872</td>
<td>87.636364</td>
<td>7.352941</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1334</td>
<td>ENSMUST00000066379.10</td>
<td>8.227374</td>
<td>9.537246</td>
<td>1.913624e-16</td>
<td>4.768176e-14</td>
<td>radial glia</td>
<td>imature Glutamatergic</td>
<td>imature Glutamatergic__radial glia</td>
<td>Chchd3</td>
<td>68</td>
<td>234</td>
<td>302</td>
<td>2.459492e-12</td>
<td>-9.537246</td>
<td>66.176471</td>
<td>0.854701</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1338</td>
<td>ENSMUST00000066379.10</td>
<td>8.379933</td>
<td>9.760194</td>
<td>5.295773e-17</td>
<td>1.090535e-14</td>
<td>radial glia</td>
<td>mature Glutamatergic</td>
<td>mature Glutamatergic__radial glia</td>
<td>Chchd3</td>
<td>68</td>
<td>275</td>
<td>343</td>
<td>6.106998e-13</td>
<td>-9.760194</td>
<td>66.176471</td>
<td>0.727273</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2567</td>
<td>ENSMUST00000035038.7</td>
<td>4.694975</td>
<td>32.364899</td>
<td>2.666390e-06</td>
<td>1.221295e-04</td>
<td>radial glia</td>
<td>mature Glutamatergic</td>
<td>mature Glutamatergic__radial glia</td>
<td>Faim</td>
<td>68</td>
<td>275</td>
<td>343</td>
<td>1.323688e-03</td>
<td>-32.364899</td>
<td>36.764706</td>
<td>0.000000</td>
</tr>
</tbody>
</table>

<p>124 rows Ã— 16 columns</p>
</div>
</div>
</div>
<div id="86056d77" class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>dirichlet_results</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">gene_id</th>
<th data-quarto-table-cell-role="th">transcript_id</th>
<th data-quarto-table-cell-role="th">group_1</th>
<th data-quarto-table-cell-role="th">group_2</th>
<th data-quarto-table-cell-role="th">contrast</th>
<th data-quarto-table-cell-role="th">p_value</th>
<th data-quarto-table-cell-role="th">chi2_stat</th>
<th data-quarto-table-cell-role="th">log_fold_change</th>
<th data-quarto-table-cell-role="th">avg_usage_group_1</th>
<th data-quarto-table-cell-role="th">avg_usage_group_2</th>
<th data-quarto-table-cell-role="th">n_cells_group_1</th>
<th data-quarto-table-cell-role="th">n_cells_group_2</th>
<th data-quarto-table-cell-role="th">total_cells</th>
<th data-quarto-table-cell-role="th">percent_expressed_group_1</th>
<th data-quarto-table-cell-role="th">percent_expressed_group_2</th>
<th data-quarto-table-cell-role="th">adj_pval</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">3</td>
<td>Clta</td>
<td>ENSMUST00000170241.7</td>
<td>mature Glutamatergic</td>
<td>intermediate progenitor</td>
<td>mature Glutamatergic__intermediate progenitor</td>
<td>1.050406e-07</td>
<td>32.137838</td>
<td>-1.300540</td>
<td>0.147946</td>
<td>0.543153</td>
<td>275</td>
<td>86</td>
<td>361</td>
<td>47.636364</td>
<td>73.255814</td>
<td>5.383331e-07</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4</td>
<td>Clta</td>
<td>ENSMUST00000107849.9</td>
<td>mature Glutamatergic</td>
<td>intermediate progenitor</td>
<td>mature Glutamatergic__intermediate progenitor</td>
<td>1.050406e-07</td>
<td>32.137838</td>
<td>0.723716</td>
<td>0.750236</td>
<td>0.363824</td>
<td>275</td>
<td>86</td>
<td>361</td>
<td>87.636364</td>
<td>55.813953</td>
<td>5.383331e-07</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">7</td>
<td>Pkm</td>
<td>ENSMUST00000163694.3</td>
<td>mature Glutamatergic</td>
<td>intermediate progenitor</td>
<td>mature Glutamatergic__intermediate progenitor</td>
<td>2.930989e-14</td>
<td>62.319359</td>
<td>1.289581</td>
<td>0.786460</td>
<td>0.216579</td>
<td>275</td>
<td>86</td>
<td>361</td>
<td>89.818182</td>
<td>51.162791</td>
<td>3.004264e-13</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">8</td>
<td>Pkm</td>
<td>ENSMUST00000034834.15</td>
<td>mature Glutamatergic</td>
<td>intermediate progenitor</td>
<td>mature Glutamatergic__intermediate progenitor</td>
<td>2.930989e-14</td>
<td>62.319359</td>
<td>-1.810307</td>
<td>0.126267</td>
<td>0.771793</td>
<td>275</td>
<td>86</td>
<td>361</td>
<td>42.181818</td>
<td>94.186047</td>
<td>3.004264e-13</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9</td>
<td>Srsf3</td>
<td>ENSMUST00000130216.1</td>
<td>mature Glutamatergic</td>
<td>intermediate progenitor</td>
<td>mature Glutamatergic__intermediate progenitor</td>
<td>2.658153e-02</td>
<td>7.255078</td>
<td>0.146748</td>
<td>0.881950</td>
<td>0.761573</td>
<td>275</td>
<td>86</td>
<td>361</td>
<td>92.727273</td>
<td>91.860465</td>
<td>4.541011e-02</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">10</td>
<td>Srsf3</td>
<td>ENSMUST00000037776.14</td>
<td>mature Glutamatergic</td>
<td>intermediate progenitor</td>
<td>mature Glutamatergic__intermediate progenitor</td>
<td>2.658153e-02</td>
<td>7.255078</td>
<td>-1.165232</td>
<td>0.052596</td>
<td>0.168659</td>
<td>275</td>
<td>86</td>
<td>361</td>
<td>21.090909</td>
<td>52.325581</td>
<td>4.541011e-02</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">11</td>
<td>Myl6</td>
<td>ENSMUST00000218127.1</td>
<td>mature Glutamatergic</td>
<td>radial glia</td>
<td>mature Glutamatergic__radial glia</td>
<td>1.018093e-02</td>
<td>9.174478</td>
<td>-1.448639</td>
<td>0.143715</td>
<td>0.611841</td>
<td>275</td>
<td>68</td>
<td>343</td>
<td>37.818182</td>
<td>89.705882</td>
<td>1.897355e-02</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">12</td>
<td>Myl6</td>
<td>ENSMUST00000164181.1</td>
<td>mature Glutamatergic</td>
<td>radial glia</td>
<td>mature Glutamatergic__radial glia</td>
<td>1.018093e-02</td>
<td>9.174478</td>
<td>0.741052</td>
<td>0.783558</td>
<td>0.373453</td>
<td>275</td>
<td>68</td>
<td>343</td>
<td>89.090909</td>
<td>75.000000</td>
<td>1.897355e-02</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">13</td>
<td>Tecr</td>
<td>ENSMUST00000019382.16</td>
<td>mature Glutamatergic</td>
<td>radial glia</td>
<td>mature Glutamatergic__radial glia</td>
<td>7.260738e-07</td>
<td>28.271228</td>
<td>-0.825942</td>
<td>0.339977</td>
<td>0.776519</td>
<td>275</td>
<td>68</td>
<td>343</td>
<td>67.636364</td>
<td>91.176471</td>
<td>2.480752e-06</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">14</td>
<td>Tecr</td>
<td>ENSMUST00000165740.8</td>
<td>mature Glutamatergic</td>
<td>radial glia</td>
<td>mature Glutamatergic__radial glia</td>
<td>7.260738e-07</td>
<td>28.271228</td>
<td>1.259196</td>
<td>0.580023</td>
<td>0.164658</td>
<td>275</td>
<td>68</td>
<td>343</td>
<td>83.636364</td>
<td>45.588235</td>
<td>2.480752e-06</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">15</td>
<td>Cdc42</td>
<td>ENSMUST00000030417.9</td>
<td>mature Glutamatergic</td>
<td>cycling radial glia</td>
<td>mature Glutamatergic__cycling radial glia</td>
<td>2.291634e-05</td>
<td>21.367320</td>
<td>1.167607</td>
<td>0.589775</td>
<td>0.183484</td>
<td>275</td>
<td>117</td>
<td>392</td>
<td>85.090909</td>
<td>39.316239</td>
<td>5.872313e-05</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">16</td>
<td>Cdc42</td>
<td>ENSMUST00000051477.12</td>
<td>mature Glutamatergic</td>
<td>cycling radial glia</td>
<td>mature Glutamatergic__cycling radial glia</td>
<td>2.291634e-05</td>
<td>21.367320</td>
<td>-0.730720</td>
<td>0.352043</td>
<td>0.731046</td>
<td>275</td>
<td>117</td>
<td>392</td>
<td>72.727273</td>
<td>85.470085</td>
<td>5.872313e-05</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">17</td>
<td>Clta</td>
<td>ENSMUST00000170241.7</td>
<td>mature Glutamatergic</td>
<td>cycling radial glia</td>
<td>mature Glutamatergic__cycling radial glia</td>
<td>4.306333e-04</td>
<td>15.500507</td>
<td>-1.559435</td>
<td>0.147946</td>
<td>0.703653</td>
<td>275</td>
<td>117</td>
<td>392</td>
<td>47.636364</td>
<td>83.760684</td>
<td>9.808870e-04</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">18</td>
<td>Clta</td>
<td>ENSMUST00000107849.9</td>
<td>mature Glutamatergic</td>
<td>cycling radial glia</td>
<td>mature Glutamatergic__cycling radial glia</td>
<td>4.306333e-04</td>
<td>15.500507</td>
<td>1.310486</td>
<td>0.750236</td>
<td>0.202330</td>
<td>275</td>
<td>117</td>
<td>392</td>
<td>87.636364</td>
<td>39.316239</td>
<td>9.808870e-04</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">19</td>
<td>Pkm</td>
<td>ENSMUST00000163694.3</td>
<td>mature Glutamatergic</td>
<td>cycling radial glia</td>
<td>mature Glutamatergic__cycling radial glia</td>
<td>0.000000e+00</td>
<td>84.925988</td>
<td>1.689738</td>
<td>0.786460</td>
<td>0.145154</td>
<td>275</td>
<td>117</td>
<td>392</td>
<td>89.818182</td>
<td>43.589744</td>
<td>0.000000e+00</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">20</td>
<td>Pkm</td>
<td>ENSMUST00000034834.15</td>
<td>mature Glutamatergic</td>
<td>cycling radial glia</td>
<td>mature Glutamatergic__cycling radial glia</td>
<td>0.000000e+00</td>
<td>84.925988</td>
<td>-1.829145</td>
<td>0.126267</td>
<td>0.786470</td>
<td>275</td>
<td>117</td>
<td>392</td>
<td>42.181818</td>
<td>88.888889</td>
<td>0.000000e+00</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">21</td>
<td>Pkm</td>
<td>ENSMUST00000163694.3</td>
<td>mature Glutamatergic</td>
<td>imature GABAergic</td>
<td>mature Glutamatergic__imature GABAergic</td>
<td>1.604636e-06</td>
<td>26.685227</td>
<td>0.556169</td>
<td>0.786460</td>
<td>0.450957</td>
<td>275</td>
<td>141</td>
<td>416</td>
<td>89.818182</td>
<td>65.248227</td>
<td>4.699292e-06</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">22</td>
<td>Pkm</td>
<td>ENSMUST00000034834.15</td>
<td>mature Glutamatergic</td>
<td>imature GABAergic</td>
<td>mature Glutamatergic__imature GABAergic</td>
<td>1.604636e-06</td>
<td>26.685227</td>
<td>-1.254394</td>
<td>0.126267</td>
<td>0.442660</td>
<td>275</td>
<td>141</td>
<td>416</td>
<td>42.181818</td>
<td>65.957447</td>
<td>4.699292e-06</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">25</td>
<td>Pkm</td>
<td>ENSMUST00000163694.3</td>
<td>imature Glutamatergic</td>
<td>intermediate progenitor</td>
<td>imature Glutamatergic__intermediate progenitor</td>
<td>2.830697e-07</td>
<td>30.155146</td>
<td>1.066032</td>
<td>0.628912</td>
<td>0.216579</td>
<td>234</td>
<td>86</td>
<td>320</td>
<td>84.615385</td>
<td>51.162791</td>
<td>1.160586e-06</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">26</td>
<td>Pkm</td>
<td>ENSMUST00000034834.15</td>
<td>imature Glutamatergic</td>
<td>intermediate progenitor</td>
<td>imature Glutamatergic__intermediate progenitor</td>
<td>2.830697e-07</td>
<td>30.155146</td>
<td>-1.055774</td>
<td>0.268524</td>
<td>0.771793</td>
<td>234</td>
<td>86</td>
<td>320</td>
<td>58.974359</td>
<td>94.186047</td>
<td>1.160586e-06</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">29</td>
<td>Tecr</td>
<td>ENSMUST00000019382.16</td>
<td>imature Glutamatergic</td>
<td>radial glia</td>
<td>imature Glutamatergic__radial glia</td>
<td>6.528992e-04</td>
<td>14.668176</td>
<td>-0.510518</td>
<td>0.466054</td>
<td>0.776519</td>
<td>234</td>
<td>68</td>
<td>302</td>
<td>71.794872</td>
<td>91.176471</td>
<td>1.338443e-03</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">30</td>
<td>Tecr</td>
<td>ENSMUST00000165740.8</td>
<td>imature Glutamatergic</td>
<td>radial glia</td>
<td>imature Glutamatergic__radial glia</td>
<td>6.528992e-04</td>
<td>14.668176</td>
<td>0.963121</td>
<td>0.431382</td>
<td>0.164658</td>
<td>234</td>
<td>68</td>
<td>302</td>
<td>67.948718</td>
<td>45.588235</td>
<td>1.338443e-03</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">33</td>
<td>Pkm</td>
<td>ENSMUST00000163694.3</td>
<td>imature Glutamatergic</td>
<td>cycling radial glia</td>
<td>imature Glutamatergic__cycling radial glia</td>
<td>5.309121e-09</td>
<td>38.107679</td>
<td>1.466188</td>
<td>0.628912</td>
<td>0.145154</td>
<td>234</td>
<td>117</td>
<td>351</td>
<td>84.615385</td>
<td>43.589744</td>
<td>3.627899e-08</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">34</td>
<td>Pkm</td>
<td>ENSMUST00000034834.15</td>
<td>imature Glutamatergic</td>
<td>cycling radial glia</td>
<td>imature Glutamatergic__cycling radial glia</td>
<td>5.309121e-09</td>
<td>38.107679</td>
<td>-1.074612</td>
<td>0.268524</td>
<td>0.786470</td>
<td>234</td>
<td>117</td>
<td>351</td>
<td>58.974359</td>
<td>88.888889</td>
<td>3.627899e-08</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="4e33e4de" class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>wilcoxon_results.head(<span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">transcript_id</th>
<th data-quarto-table-cell-role="th">scores</th>
<th data-quarto-table-cell-role="th">log_fold_change</th>
<th data-quarto-table-cell-role="th">pvals</th>
<th data-quarto-table-cell-role="th">pvals_adj</th>
<th data-quarto-table-cell-role="th">group_1</th>
<th data-quarto-table-cell-role="th">group_2</th>
<th data-quarto-table-cell-role="th">contrast</th>
<th data-quarto-table-cell-role="th">gene_id</th>
<th data-quarto-table-cell-role="th">n_cells_group_1</th>
<th data-quarto-table-cell-role="th">n_cells_group_2</th>
<th data-quarto-table-cell-role="th">total_cells</th>
<th data-quarto-table-cell-role="th">p_value</th>
<th data-quarto-table-cell-role="th">direction</th>
<th data-quarto-table-cell-role="th">percent_expressed_group_1</th>
<th data-quarto-table-cell-role="th">percent_expressed_group_2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">5893</td>
<td>ENSMUST00000034834.15</td>
<td>12.272116</td>
<td>5.359763</td>
<td>1.278680e-34</td>
<td>8.850244e-32</td>
<td>cycling radial glia</td>
<td>mature Glutamatergic</td>
<td>cycling radial glia__mature Glutamatergic</td>
<td>Pkm</td>
<td>117</td>
<td>275</td>
<td>392</td>
<td>5.441936e-29</td>
<td>5.359763</td>
<td>88.888889</td>
<td>42.181818</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5894</td>
<td>ENSMUST00000163694.3</td>
<td>11.959407</td>
<td>5.031865</td>
<td>5.797459e-33</td>
<td>3.523300e-30</td>
<td>mature Glutamatergic</td>
<td>cycling radial glia</td>
<td>cycling radial glia__mature Glutamatergic</td>
<td>Pkm</td>
<td>275</td>
<td>117</td>
<td>392</td>
<td>1.645196e-27</td>
<td>-5.031865</td>
<td>89.818182</td>
<td>43.589744</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">5909</td>
<td>ENSMUST00000034834.15</td>
<td>11.985371</td>
<td>5.971428</td>
<td>4.239462e-33</td>
<td>2.112694e-29</td>
<td>intermediate progenitor</td>
<td>mature Glutamatergic</td>
<td>intermediate progenitor__mature Glutamatergic</td>
<td>Pkm</td>
<td>86</td>
<td>275</td>
<td>361</td>
<td>8.865320e-27</td>
<td>5.971428</td>
<td>94.186047</td>
<td>42.181818</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1477</td>
<td>ENSMUST00000107849.9</td>
<td>11.686639</td>
<td>5.032757</td>
<td>1.491739e-31</td>
<td>7.908438e-29</td>
<td>mature Glutamatergic</td>
<td>cycling radial glia</td>
<td>cycling radial glia__mature Glutamatergic</td>
<td>Clta</td>
<td>275</td>
<td>117</td>
<td>392</td>
<td>2.994790e-26</td>
<td>-5.032757</td>
<td>87.636364</td>
<td>39.316239</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">5916</td>
<td>ENSMUST00000034834.15</td>
<td>11.390181</td>
<td>6.393306</td>
<td>4.680185e-30</td>
<td>4.319117e-27</td>
<td>radial glia</td>
<td>mature Glutamatergic</td>
<td>mature Glutamatergic__radial glia</td>
<td>Pkm</td>
<td>68</td>
<td>275</td>
<td>343</td>
<td>1.156183e-24</td>
<td>-6.393306</td>
<td>95.588235</td>
<td>42.181818</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5915</td>
<td>ENSMUST00000163694.3</td>
<td>11.155944</td>
<td>8.153163</td>
<td>6.697808e-29</td>
<td>4.768266e-26</td>
<td>mature Glutamatergic</td>
<td>radial glia</td>
<td>mature Glutamatergic__radial glia</td>
<td>Pkm</td>
<td>275</td>
<td>68</td>
<td>343</td>
<td>1.039047e-23</td>
<td>8.153163</td>
<td>89.818182</td>
<td>16.176471</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1497</td>
<td>ENSMUST00000107849.9</td>
<td>10.725714</td>
<td>8.955872</td>
<td>7.708972e-27</td>
<td>4.268543e-24</td>
<td>mature Glutamatergic</td>
<td>radial glia</td>
<td>mature Glutamatergic__radial glia</td>
<td>Clta</td>
<td>275</td>
<td>68</td>
<td>343</td>
<td>7.447796e-22</td>
<td>8.955872</td>
<td>87.636364</td>
<td>7.352941</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2220</td>
<td>ENSMUST00000154584.8</td>
<td>10.593913</td>
<td>6.469005</td>
<td>3.180146e-26</td>
<td>1.576562e-23</td>
<td>radial glia</td>
<td>mature Glutamatergic</td>
<td>mature Glutamatergic__radial glia</td>
<td>Eef1d</td>
<td>68</td>
<td>275</td>
<td>343</td>
<td>2.569838e-21</td>
<td>-6.469005</td>
<td>89.705882</td>
<td>24.727273</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">5890</td>
<td>ENSMUST00000163694.3</td>
<td>10.557364</td>
<td>4.570864</td>
<td>4.696683e-26</td>
<td>2.600606e-23</td>
<td>imature Glutamatergic</td>
<td>cycling radial glia</td>
<td>cycling radial glia__imature Glutamatergic</td>
<td>Pkm</td>
<td>234</td>
<td>117</td>
<td>351</td>
<td>4.162578e-21</td>
<td>-4.570864</td>
<td>84.615385</td>
<td>43.589744</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1117</td>
<td>ENSMUST00000030417.9</td>
<td>10.449820</td>
<td>7.493307</td>
<td>1.468039e-25</td>
<td>6.773914e-23</td>
<td>mature Glutamatergic</td>
<td>radial glia</td>
<td>mature Glutamatergic__radial glia</td>
<td>Cdc42</td>
<td>275</td>
<td>68</td>
<td>343</td>
<td>1.016153e-20</td>
<td>7.493307</td>
<td>85.090909</td>
<td>14.705882</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1105</td>
<td>ENSMUST00000030417.9</td>
<td>10.376866</td>
<td>4.580756</td>
<td>3.159765e-25</td>
<td>8.846278e-23</td>
<td>mature Glutamatergic</td>
<td>cycling radial glia</td>
<td>cycling radial glia__mature Glutamatergic</td>
<td>Cdc42</td>
<td>275</td>
<td>117</td>
<td>392</td>
<td>1.314328e-20</td>
<td>-4.580756</td>
<td>85.090909</td>
<td>39.316239</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5905</td>
<td>ENSMUST00000163694.3</td>
<td>10.207308</td>
<td>7.692162</td>
<td>1.838968e-24</td>
<td>1.018257e-21</td>
<td>imature Glutamatergic</td>
<td>radial glia</td>
<td>imature Glutamatergic__radial glia</td>
<td>Pkm</td>
<td>234</td>
<td>68</td>
<td>302</td>
<td>1.320205e-19</td>
<td>7.692162</td>
<td>84.615385</td>
<td>16.176471</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1475</td>
<td>ENSMUST00000107849.9</td>
<td>10.059887</td>
<td>4.801264</td>
<td>8.309374e-24</td>
<td>5.049870e-21</td>
<td>mature GABAergic</td>
<td>cycling radial glia</td>
<td>cycling radial glia__mature GABAergic</td>
<td>Clta</td>
<td>172</td>
<td>117</td>
<td>289</td>
<td>5.906161e-19</td>
<td>-4.801264</td>
<td>83.139535</td>
<td>39.316239</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1473</td>
<td>ENSMUST00000107849.9</td>
<td>9.974315</td>
<td>4.474297</td>
<td>1.974594e-23</td>
<td>7.809675e-21</td>
<td>imature Glutamatergic</td>
<td>cycling radial glia</td>
<td>cycling radial glia__imature Glutamatergic</td>
<td>Clta</td>
<td>234</td>
<td>117</td>
<td>351</td>
<td>8.932083e-19</td>
<td>-4.474297</td>
<td>83.760684</td>
<td>39.316239</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1488</td>
<td>ENSMUST00000107849.9</td>
<td>9.956464</td>
<td>8.397410</td>
<td>2.363200e-23</td>
<td>1.154585e-20</td>
<td>imature Glutamatergic</td>
<td>radial glia</td>
<td>imature Glutamatergic__radial glia</td>
<td>Clta</td>
<td>234</td>
<td>68</td>
<td>302</td>
<td>1.289647e-18</td>
<td>8.397410</td>
<td>83.760684</td>
<td>7.352941</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">175</td>
<td>ENSMUST00000044492.9</td>
<td>9.735595</td>
<td>6.681633</td>
<td>2.125714e-22</td>
<td>9.808594e-20</td>
<td>imature Glutamatergic</td>
<td>radial glia</td>
<td>imature Glutamatergic__radial glia</td>
<td>Akap9</td>
<td>234</td>
<td>68</td>
<td>302</td>
<td>9.840919e-18</td>
<td>6.681633</td>
<td>83.760684</td>
<td>22.058824</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">168</td>
<td>ENSMUST00000044492.9</td>
<td>9.685860</td>
<td>3.796590</td>
<td>3.462841e-22</td>
<td>1.250487e-19</td>
<td>imature Glutamatergic</td>
<td>cycling radial glia</td>
<td>cycling radial glia__imature Glutamatergic</td>
<td>Akap9</td>
<td>234</td>
<td>117</td>
<td>351</td>
<td>1.232702e-17</td>
<td>-3.796590</td>
<td>83.760684</td>
<td>50.427350</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">6923</td>
<td>ENSMUST00000102843.9</td>
<td>9.674722</td>
<td>4.648714</td>
<td>3.861395e-22</td>
<td>1.374491e-19</td>
<td>mature Glutamatergic</td>
<td>radial glia</td>
<td>mature Glutamatergic__radial glia</td>
<td>Rtn4</td>
<td>275</td>
<td>68</td>
<td>343</td>
<td>1.342160e-17</td>
<td>4.648714</td>
<td>88.000000</td>
<td>47.058824</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1495</td>
<td>ENSMUST00000107849.9</td>
<td>9.634603</td>
<td>8.724378</td>
<td>5.711261e-22</td>
<td>2.846150e-19</td>
<td>mature GABAergic</td>
<td>radial glia</td>
<td>mature GABAergic__radial glia</td>
<td>Clta</td>
<td>172</td>
<td>68</td>
<td>240</td>
<td>2.694471e-17</td>
<td>8.724378</td>
<td>83.139535</td>
<td>7.352941</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5914</td>
<td>ENSMUST00000034834.15</td>
<td>9.559292</td>
<td>5.820885</td>
<td>1.185652e-21</td>
<td>5.681323e-19</td>
<td>radial glia</td>
<td>mature GABAergic</td>
<td>mature GABAergic__radial glia</td>
<td>Pkm</td>
<td>68</td>
<td>172</td>
<td>240</td>
<td>5.196361e-17</td>
<td>-5.820885</td>
<td>95.588235</td>
<td>43.023256</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="54523e7b" class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> allos.visuals <span class="im">as</span> vs</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(combined_mouse_data, target_sum<span class="op">=</span><span class="fl">1e6</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(combined_mouse_data)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(combined_mouse_data)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(combined_mouse_data)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(combined_mouse_data, color <span class="op">=</span> <span class="st">'cell_type'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/scanpy/tools/_utils.py:41: UserWarning: Youâ€™re trying to run this on 24917 dimensions of `.X`, if you really want this, set `use_rep='X'`.
         Falling back to preprocessing with `sc.pp.pca` and default params.
  warnings.warn(</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="005_switch_search_files/figure-html/cell-15-output-2.png" width="758" height="430" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="9ac10260" class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>vs.plot_transcripts(combined_mouse_data, gene_id<span class="op">=</span><span class="st">'Pkm'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="005_switch_search_files/figure-html/cell-16-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="f42f2549" class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>Pkm_Txs <span class="op">=</span> combined_mouse_data[:, combined_mouse_data.var[<span class="st">'geneId'</span>] <span class="op">==</span><span class="st">'Pkm'</span>].var.index.to_list()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="978ba452" class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>vs.plot_density_multi(combined_mouse_data, features<span class="op">=</span>Pkm_Txs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="005_switch_search_files/figure-html/cell-18-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="4a749038" class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>vs.plot_transcripts(combined_mouse_data, gene_id<span class="op">=</span><span class="st">'Cdc42'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="005_switch_search_files/figure-html/cell-19-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="a4a369d1" class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>Eef1d_Txs <span class="op">=</span> combined_mouse_data[:, combined_mouse_data.var[<span class="st">'geneId'</span>] <span class="op">==</span><span class="st">'Cdc42'</span>].var.index.to_list()</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>vs.plot_density_multi(combined_mouse_data, features<span class="op">=</span>Eef1d_Txs)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="005_switch_search_files/figure-html/cell-20-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="442e388a" class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>vs.plot_transcript_exspression_dotplot(combined_mouse_data, gene_id<span class="op">=</span> <span class="st">'Cdc42'</span>, top_n<span class="op">=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="005_switch_search_files/figure-html/cell-21-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="7b1efde4" class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>vs.plot_transcript_expression_violin(combined_mouse_data, gene_id<span class="op">=</span> <span class="st">'Cdc42'</span>, top_n<span class="op">=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="005_switch_search_files/figure-html/cell-22-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="bb78c7d3" class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> urllib.request</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> pathlib <span class="im">import</span> Path</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Example Ensembl URLs for mouse GRCm39 (release 109)</span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>gtf_url <span class="op">=</span> <span class="st">"ftp://ftp.ensembl.org/pub/release-109/gtf/mus_musculus/Mus_musculus.GRCm39.109.gtf.gz"</span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>fasta_url <span class="op">=</span> <span class="st">"ftp://ftp.ensembl.org/pub/release-109/fasta/mus_musculus/dna/Mus_musculus.GRCm39.dna.primary_assembly.fa.gz"</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Store data one directory back</span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>data_dir <span class="op">=</span> Path(<span class="st">".."</span>) <span class="op">/</span> <span class="st">"data"</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>data_dir.mkdir(parents<span class="op">=</span><span class="va">True</span>, exist_ok<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>gtf_file_local <span class="op">=</span> data_dir <span class="op">/</span> <span class="st">"Mus_musculus.GRCm39.109.gtf.gz"</span></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>fasta_file_local <span class="op">=</span> data_dir <span class="op">/</span> <span class="st">"Mus_musculus.GRCm39.dna.primary_assembly.fa.gz"</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Download if not already present</span></span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> gtf_file_local.is_file():</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Downloading </span><span class="sc">{</span>gtf_url<span class="sc">}</span><span class="ss">..."</span>)</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>    urllib.request.urlretrieve(gtf_url, gtf_file_local)</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> fasta_file_local.is_file():</span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Downloading </span><span class="sc">{</span>fasta_url<span class="sc">}</span><span class="ss">..."</span>)</span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>    urllib.request.urlretrieve(fasta_url, fasta_file_local)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<p><a href="https://github.com/cobioda/allos/blob/main/allos/switch_search.py#L710" target="_blank" style="float:right; font-size:smaller">source</a></p>
</section>
<section id="get_top_n_isoforms" class="level3">
<h3 class="anchored" data-anchor-id="get_top_n_isoforms">get_top_n_isoforms</h3>
<blockquote class="blockquote">
<pre><code> get_top_n_isoforms (adata, gene_id, top_n=5, strip=False)</code></pre>
</blockquote>
<p>*Get the top_n transcript IDs for a given gene based on overall average expression.</p>
<p>Parameters: - adata: AnnData object containing transcript-level data. - gene_id: The gene ID to search for corresponding isoforms. - top_n: The number of top isoforms to return based on overall average expression. - strip: Boolean indicating whether to strip the dot and everything after from the ID.</p>
<p>Returns: - A list of top_n transcript IDs.*</p>
<div id="cafba89b" class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>top_n <span class="op">=</span> get_top_n_isoforms(combined_mouse_data,  gene_id<span class="op">=</span> <span class="st">'Cdc42'</span>, top_n<span class="op">=</span><span class="dv">2</span>, strip <span class="op">=</span> <span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="4dc32719" class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>top_n</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>['ENSMUST00000051477', 'ENSMUST00000030417']</code></pre>
</div>
</div>
<div id="0325fbc7" class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> allos.color_palette <span class="im">import</span> ghibli</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> allos.transcript_plots <span class="im">import</span> TranscriptPlots</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>tp <span class="op">=</span> TranscriptPlots(gtf_file<span class="op">=</span>gtf_file_local)</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>tp.draw_transcripts_list([<span class="st">'ENSMUST00000051477'</span>, <span class="st">'ENSMUST00000030417'</span>], colors<span class="op">=</span>ghibli,  draw_cds<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="005_switch_search_files/figure-html/cell-27-output-1.png" width="1570" height="328" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="9eb8402a" class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>combined_mouse_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>AnnData object with n_obs Ã— n_vars = 1109 Ã— 24917
    obs: 'batch', 'cell_type'
    var: 'geneId'
    uns: 'log1p', 'pca', 'neighbors', 'umap', 'cell_type_colors'
    obsm: 'X_pca', 'X_umap'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'</code></pre>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "î§‹";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/cobioda\.github\.io\/allos");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/cobioda/allos/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>