{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56daa03d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata as ad\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import requests\n",
    "import sys\n",
    "import patchworklib as pw\n",
    "import urllib \n",
    "import os\n",
    "import gzip\n",
    "import shutil\n",
    "\n",
    "import torch\n",
    "import pyro\n",
    "import pyro.distributions as dist\n",
    "from pyro.infer import SVI, Trace_ELBO\n",
    "from pyro.optim import Adam\n",
    "from joblib import Parallel, delayed\n",
    "from itertools import combinations\n",
    "from threading import Thread"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "742d1917",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File downloaded successfully\n",
      "File extracted successfully\n"
     ]
    }
   ],
   "source": [
    "def download_test_data() -> str: #The absolute path of the extracted file 'sample_isomatrix.txt' if the download is successful.\n",
    "    \"\"\"\n",
    "    This function downloads a test data file from a specified URL, saves it locally, and extracts it.\n",
    "    \"\"\"\n",
    "\n",
    "    # URL of the file to be downloaded\n",
    "    url = \"https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3748nnn/GSM3748087/suppl/GSM3748087%5F190c.isoforms.matrix.txt.gz\"\n",
    "\n",
    "    # Download the file from `url` and save it locally under `file.txt.gz`:\n",
    "    urllib.request.urlretrieve(url, 'file.txt.gz')\n",
    "\n",
    "    # Check if the file is downloaded correctly\n",
    "    if os.path.exists('file.txt.gz'):\n",
    "        print(\"File downloaded successfully\")\n",
    "        # Now we need to extract the file\n",
    "        with gzip.open('file.txt.gz', 'rb') as f_in:\n",
    "            with open('sample_isomatrix.txt', 'wb') as f_out:\n",
    "                shutil.copyfileobj(f_in, f_out)\n",
    "        print(\"File extracted successfully\")\n",
    "        return os.path.abspath('sample_isomatrix.txt')\n",
    "    else:\n",
    "        print(\"Failed to download the file\")\n",
    "        return None \n",
    "    \n",
    "path = download_test_data()\n",
    "\n",
    "def create_iso_adata (path: str # path to a tab separated file, sicelore output with transcript counts and columns transcriptId, geneId\n",
    "):\n",
    "\n",
    "    \n",
    "    with open(path) as isoforms:\n",
    "        df = pd.read_table(isoforms, sep='\\t')\n",
    "\n",
    "    counts = df.iloc[0:,2:]\n",
    "    adata_iso = ad.AnnData(counts).transpose()\n",
    "    adata_iso.var_names = df['transcriptId'].to_list()\n",
    "    adata_iso.var = df[['transcriptId', 'geneId']]\n",
    "    barcodes = {'barcodes': df.columns.values[2:]}   \n",
    "    barcodes = pd.DataFrame(data=barcodes)\n",
    "    adata_iso.obs = barcodes\n",
    "    adata_iso.obs_names = barcodes['barcodes'].tolist()\n",
    "    return(adata_iso)\n",
    "\n",
    "adata_iso = create_iso_adata(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "391cd753",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_type = pd.read_csv('./annot_190c.csv', sep='\\t')\n",
    "adata_iso.obs['cell_type'] = cell_type['leiden'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7360aed0",
   "metadata": {},
   "outputs": [],
   "source": [
    "class AnnDataIso(ad.AnnData):\n",
    "    \n",
    "    def __filter_isodata (self):\n",
    "    #     ##\n",
    "        genes, frequencies = np.unique(self.var['geneId'], return_counts=True)\n",
    "        df = pd.DataFrame({\"geneId\": genes, \"frequencies\": frequencies})\n",
    "        multi_iso_genes = df[df[\"frequencies\"]>1]['geneId'].tolist()\n",
    "        filtr_adata_iso = self[:, self.var['geneId'].isin(multi_iso_genes)]\n",
    "        return filtr_adata_iso\n",
    "    \n",
    "    \n",
    "    def iso_percent(self, df, barcodes_regex=\"^[ACGT]+$\"):\n",
    "        iso_perc_df = df.__deepcopy__()\n",
    "        if len(iso_perc_df.filter(regex=(barcodes_regex)).columns.to_list()) < 1:\n",
    "            raise ValueError(\"No barcode was identified. Please check the names of the columns.\")\n",
    "        iso_perc_df[iso_perc_df.filter(regex=(barcodes_regex)).columns.to_list()] = iso_perc_df.filter(regex=(barcodes_regex)) / iso_perc_df.groupby(['geneId']).transform('sum').filter(regex=(barcodes_regex))\n",
    "        iso_perc_df = iso_perc_df.replace(np.nan, 0.0)\n",
    "        return iso_perc_df\n",
    "        \n",
    "    def __init__(self, anndata: ad.AnnData, cell_types: pd.DataFrame):\n",
    "        self._init_as_actual(anndata.copy())\n",
    "        self.gene_counts = self.var.reset_index().groupby(by='geneId').count()\n",
    "        self.__filtered_anndata = self.__filter_isodata()\n",
    "        self.obs['cell_type'] = cell_types\n",
    "        df = self.__filtered_anndata.to_df().set_index(self.__filtered_anndata.obs['barcodes'])\n",
    "        df = df.transpose()\n",
    "        df[['transcriptId', 'geneId']] = self.__filtered_anndata.var\n",
    "        df_m_iso = self.iso_percent(df)\n",
    "        df_m_iso = df_m_iso.iloc[0:,:-2].transpose()\n",
    "        self.__filtered_anndata.obsm['Iso_prct'] = df_m_iso\n",
    "\n",
    "    def plot_isoforms_summary(self):\n",
    "        ax1 = pw.Brick(figsize=(4,4))\n",
    "        self._plot_switch_gen_bar(ax1)\n",
    "        ax1.set_title(\"Multiple isoforms genes %\")\n",
    "        ax2 = pw.Brick(figsize=(4,4))\n",
    "        self._plot_isoforms_frequencies(ax2)\n",
    "        ax2.set_title(\"Frequency of isoforms per gene\")\n",
    "        ax3 = pw.Brick(figsize=(3,2))\n",
    "        self._plot_genes_cell_type(ax3)\n",
    "        ax3.set_title(\"Nb of genes per cell type\")\n",
    "        return (ax1|ax2|ax3).savefig()\n",
    "\n",
    "    def _plot_genes_cell_type (self, _ax): #mdata with annotated cell types in isoform and gen anndata objects\n",
    "        if _ax is None:\n",
    "            df = pd.DataFrame(np.transpose(self.X), columns=self.obs['cell_type'])\n",
    "            df = df.sum(axis = 0).to_frame().reset_index()\n",
    "            df.columns = ['cell_type', 'n_of_genes']\n",
    "            ax = sns.boxplot(x='cell_type', y='n_of_genes', data=df) \n",
    "            ax1 = sns.stripplot(x='cell_type', y='n_of_genes', data=df, color = 'black', size = 3)\n",
    "            ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "            plt.show()\n",
    "        else:\n",
    "            df = pd.DataFrame(np.transpose(self.X), columns=self.obs['cell_type'])\n",
    "            df = df.sum(axis = 0).to_frame().reset_index()\n",
    "            df.columns = ['cell_type', 'n_of_genes']\n",
    "            sns.boxplot(x='cell_type', y='n_of_genes', data=df, ax=_ax) \n",
    "            sns.stripplot(x='cell_type', y='n_of_genes', data=df, color = 'black', size = 3, ax=_ax)\n",
    "            _ax.set_xticklabels(_ax.get_xticklabels(), rotation=90)\n",
    "\n",
    "    def plot_genes_cell_type(self):\n",
    "        self._plot_genes_cell_type(None)\n",
    "    \n",
    "    def _plot_isoforms_frequencies (self, _ax):\n",
    "        if _ax is None:\n",
    "            fig, ax = plt.subplots()\n",
    "            self.gene_counts['transcriptId'].value_counts().plot(ax=ax, kind='bar', \n",
    "                                                                xlabel='number of isoforms per gene', ylabel='quantity of genes')\n",
    "        else:\n",
    "            self.gene_counts['transcriptId'].value_counts().plot(ax=_ax, kind='bar', \n",
    "                                                             xlabel='number of isoforms per gene', ylabel='quantity of genes')\n",
    "    def plot_isoforms_frequencies(self):\n",
    "        self._plot_isoforms_frequencies(None)\n",
    "\n",
    "    def _plot_switch_gen_bar (self, _ax): \n",
    "        if _ax is None:\n",
    "            iso_per_gene = self.gene_counts\n",
    "            x = ['genes']\n",
    "            x1 = ['transcripts']\n",
    "            multiple_iso = sum(iso_per_gene['transcriptId'] > 1)\n",
    "            mono_iso = sum((iso_per_gene['transcriptId'] > 1) == False)\n",
    "            labels = [str(round(1000*multiple_iso/(multiple_iso+mono_iso))/10) + '%', str(round(1000*mono_iso/(multiple_iso+mono_iso))/10) + '%']\n",
    "            fig, ax = plt.subplots()\n",
    "            mult = ax.bar(x, multiple_iso, color = 'deepskyblue', label=labels[1]) #multiple_iso/(multiple_iso+mono_iso))\n",
    "            mono = ax.bar(x, mono_iso, bottom=multiple_iso, color='sandybrown', label=labels[0]) #mona_iso/(multiple_iso+mono_iso))\n",
    "            tran = ax.bar(x1, len(self.var['transcriptId']))\n",
    "            ax.text(\n",
    "                ax.patches[0].get_x() + ax.patches[0].get_width() / 2, ax.patches[0].get_height() / 2, labels[0], ha=\"center\", va=\"center\"\n",
    "                )\n",
    "            ax.text(\n",
    "                ax.patches[1].get_x() + ax.patches[1].get_width() / 2, ax.patches[1].get_height() / 2 + ax.patches[0].get_height(), labels[1], ha=\"center\", va=\"center\"\n",
    "                )\n",
    "            plt.legend(['Multiple isoforms', 'Single isoform'])\n",
    "            plt.show()\n",
    "        else:\n",
    "            iso_per_gene = self.gene_counts\n",
    "            x = ['genes']\n",
    "            x1 = ['transcripts']\n",
    "            multiple_iso = sum(iso_per_gene['transcriptId'] > 1)\n",
    "            mono_iso = sum((iso_per_gene['transcriptId'] > 1) == False)\n",
    "            labels = [str(round(1000*multiple_iso/(multiple_iso+mono_iso))/10) + '%', str(round(1000*mono_iso/(multiple_iso+mono_iso))/10) + '%']\n",
    "            mult = _ax.bar(x, multiple_iso, color = 'deepskyblue', label=labels[1]) #multiple_iso/(multiple_iso+mono_iso))\n",
    "            mono = _ax.bar(x, mono_iso, bottom=multiple_iso, color='sandybrown', label=labels[0]) #mona_iso/(multiple_iso+mono_iso))\n",
    "            tran = _ax.bar(x1, len(self.var['transcriptId']))\n",
    "            _ax.text(\n",
    "                _ax.patches[0].get_x() + _ax.patches[0].get_width() / 2, _ax.patches[0].get_height() / 2, labels[0], ha=\"center\", va=\"center\"\n",
    "                )\n",
    "            _ax.text(\n",
    "                _ax.patches[1].get_x() + _ax.patches[1].get_width() / 2, _ax.patches[1].get_height() / 2 + _ax.patches[0].get_height(), labels[1], ha=\"center\", va=\"center\"\n",
    "                )\n",
    "            _ax.legend(['Multiple isoforms', 'Single isoform'])\n",
    "            plt.show()\n",
    "    \n",
    "    def plot_switch_gen_bar (self): # takes as input mdata object\n",
    "        self._plot_switch_gen_bar (None)\n",
    "\n",
    "    def _plot_transcripts_per_cell_type(self, gene_name, _ax):\n",
    "        if _ax is None:\n",
    "            grouped = self.__filtered_anndata.obsm['Iso_prct']\n",
    "            grouped['cell_type'] = self.obs['cell_type']\n",
    "            res = grouped.groupby('cell_type').mean().transpose()\n",
    "            res = res.assign(transcriptId=self.__filtered_anndata.var['transcriptId'].to_list())\n",
    "            res = res.assign(geneId=self.__filtered_anndata.var['geneId'].to_list())\n",
    "            res = res[res['geneId'] == gene_name].drop(['geneId'], axis=1)\n",
    "            plot_data = res.set_index('transcriptId').transpose()\n",
    "            plot_data.plot(kind='barh', stacked=True).legend(loc='center left',bbox_to_anchor=(1.0, 1.0))\n",
    "        else:\n",
    "            grouped = self.__filtered_anndata.obsm['Iso_prct']\n",
    "            grouped['cell_type'] = self.obs['cell_type']\n",
    "            res = grouped.groupby('cell_type').mean().transpose()\n",
    "            res = res.assign(transcriptId=self.__filtered_anndata.var['transcriptId'].to_list())\n",
    "            res = res.assign(geneId=self.__filtered_anndata.var['geneId'].to_list())\n",
    "            res = res[res['geneId'] == gene_name].drop(['geneId'], axis=1)\n",
    "            plot_data = res.set_index('transcriptId').transpose()\n",
    "            plot_data.plot(kind='barh', stacked=True, ax=_ax).legend(loc='center left',bbox_to_anchor=(1.0, 1.0))\n",
    "            \n",
    "    def plot_transcripts_per_cell_type(self, gene_name):\n",
    "        self._plot_transcripts_per_cell_type(self, gene_name, None)\n",
    "    \n",
    "    def _trsct_counts_cell_type (self, gene_name, _ax):\n",
    "        if _ax is None:\n",
    "        # create df with filtered isoform counts and labeled cell types:\n",
    "            df = self.__filtered_anndata.to_df().set_index(self.__filtered_anndata.obs['cell_type'])\n",
    "            df = df.transpose() \n",
    "            df[['transcriptId', 'geneId']] = self.__filtered_anndata.var\n",
    "            gene_iso_count = df[df['geneId']== gene_name]\n",
    "            gene_iso_count = gene_iso_count.drop('geneId', axis=1).set_index('transcriptId').transpose()\n",
    "            gene_iso_count_long = gene_iso_count.reset_index().melt(id_vars='cell_type', var_name='transcriptId', value_name='count')\n",
    "            g = sns.catplot(x=\"cell_type\", y=\"count\", col=\"transcriptId\", aspect=1, dodge=False, kind=\"violin\", data=gene_iso_count_long)\n",
    "            # Set custom facet titles\n",
    "            g.set_titles(col_template=\"{col_name}\", size = 8)\n",
    "            # Remove x ticks\n",
    "            g.set_xticklabels(rotation=90)\n",
    "            g.fig.suptitle(gene_name)\n",
    "            plt.show()\n",
    "        else:\n",
    "            df = self.__filtered_anndata.to_df().set_index(self.__filtered_anndata.obs['cell_type'])\n",
    "            df = df.transpose() \n",
    "            df[['transcriptId', 'geneId']] = self.__filtered_anndata.var\n",
    "            gene_iso_count = df[df['geneId']== gene_name]\n",
    "            gene_iso_count = gene_iso_count.drop('geneId', axis=1).set_index('transcriptId').transpose()\n",
    "            gene_iso_count_long = gene_iso_count.reset_index().melt(id_vars='cell_type', var_name='transcriptId', value_name='count')\n",
    "            g = sns.catplot(x=\"cell_type\", y=\"count\", col=\"transcriptId\", aspect=1, dodge=False, kind=\"violin\", data=gene_iso_count_long)\n",
    "            # Set custom facet titles\n",
    "            g.set_titles(col_template=\"{col_name}\", size = 8)\n",
    "            # Remove x ticks\n",
    "            g.set_xticklabels(rotation=90, labels=self.obs['cell_type'].unique())\n",
    "            g.fig.suptitle(gene_name)\n",
    "            return g\n",
    "\n",
    "    \n",
    "    def trsct_counts_cell_type (self, gene_name):\n",
    "        # create df with filtered isoform counts and labeled cell types:\n",
    "        self._trsct_counts_cell_type(gene_name, None)\n",
    "\n",
    "    def __get_coord_from_tscrpt_id(self, transcript_id):\n",
    "        if '.' in transcript_id:\n",
    "            transcript_id = transcript_id.split('.')[0]\n",
    "        server = \"https://rest.ensembl.org\"\n",
    "        ext = \"/lookup/id/\" + transcript_id + \"?expand=1\"\n",
    "    \n",
    "        r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    \n",
    "        if not r.ok:\n",
    "            r.raise_for_status()\n",
    "            sys.exit()\n",
    "    \n",
    "        decoded = r.json()\n",
    "        exon_list = list(decoded['Exon'])\n",
    "        exon_coord = []\n",
    "        for i, e in enumerate(exon_list):\n",
    "            coord = [e.get('end'), e.get('start')]\n",
    "            exon_coord.append(coord)\n",
    "        strand = decoded['strand']\n",
    "        return(exon_coord, strand)\n",
    "\n",
    "    def __draw_exons(self, exons, direction, color, transcript_name, offset=0, start_override=None, end_override=None, no_render=False):\n",
    "        if not no_render:\n",
    "            plt.axes()\n",
    "            plt.xlim((-0.1, 1))\n",
    "            plt.ylim((-0.3, 0.3))\n",
    "            plt.margins(0.2)\n",
    "            plt.axis('off')\n",
    "            fig = plt.gcf()\n",
    "            fig.set_size_inches(20, 2)\n",
    "        height = 0.2\n",
    "        plt.plot([offset + 0.1, offset + 0.1], linestyle='solid', linewidth=0.5, c='grey')\n",
    "        j = 0\n",
    "        k = 1\n",
    "        if direction == 1:\n",
    "            pos_start = exons[0][1]\n",
    "            pos_end = exons[-1][0]\n",
    "        else: #direction == -1\n",
    "            pos_start = exons[-1][1]\n",
    "            pos_end = exons[0][0]\n",
    "            j = 1\n",
    "            k = 0\n",
    "        real_start = pos_start\n",
    "        real_end = pos_end\n",
    "        if start_override is not None and end_override is not None:\n",
    "            pos_start = start_override\n",
    "            pos_end = end_override\n",
    "        total_length = pos_end - pos_start\n",
    "        total_length_with_margin = 1.05 * total_length\n",
    "        pos_start_with_margin = pos_start - 0.025*total_length\n",
    "        for i, exon in enumerate(exons):\n",
    "            rectangle = plt.Rectangle(((exon[j] - pos_start_with_margin)/total_length_with_margin,offset), (exon[k] - exon[j])/total_length_with_margin, height, fc=color,ec=\"black\")\n",
    "            plt.gca().add_patch(rectangle)\n",
    "        if i > 0:\n",
    "            arrow = None\n",
    "            if direction < 0:\n",
    "                arrow = plt.arrow(1, offset - height/4, -1, 0, width=0.0015, head_length=0.01, head_width=0.1, length_includes_head=True, overhang=1)\n",
    "            else:\n",
    "                arrow = plt.arrow(0, offset - height/4, 1, 0, width=0.0015, head_length=0.01, head_width=0.1, length_includes_head=True, overhang=1)\n",
    "            plt.gca().add_patch(arrow)\n",
    "        plt.plot(np.array([0.025 + (real_start - pos_start) / (total_length)/1.05, 0.025 + (real_start - pos_start) / (total_length)/1.05]), np.array([offset - height/4 - 0.03, offset - height/4 + 0.03]), color='black')\n",
    "        plt.plot(np.array([1 - 0.025 - (pos_end - real_end) / (total_length)/1.05, 1 - 0.025 - (pos_end - real_end) / (total_length)/1.05]), np.array([offset - height/4 - 0.03, offset - height/4 + 0.03]), color='black')\n",
    "        plt.text(0.025 + (real_start - pos_start) / (total_length)/1.05, offset - height/4 - 0.075, real_start, horizontalalignment='center', verticalalignment='center', fontsize=9)\n",
    "        plt.text(1 - 0.025 - (pos_end - real_end) / (total_length)/1.05, offset - height/4 - 0.075, real_end, horizontalalignment='center', verticalalignment='center', fontsize=9)\n",
    "        plt.text(1, offset - height, transcript_name, horizontalalignment='right', verticalalignment='top', fontsize=12)\n",
    "        if not no_render:\n",
    "            plt.show()\n",
    "    \n",
    "    def __get_transcripts_from_gene(self, gene_name):\n",
    "        elems = self.__filtered_anndata.var\n",
    "        return elems[elems['geneId'] == gene_name]['transcriptId'].to_list()\n",
    "\n",
    "    def __draw_transcripts_list(self, gene_name, _ax, colors=None):\n",
    "        transcripts_id = self.__get_transcripts_from_gene(gene_name)\n",
    "        exons = []\n",
    "        directions = []\n",
    "        for tr in transcripts_id:\n",
    "            t, d = self.__get_coord_from_tscrpt_id(tr)\n",
    "            exons += [t]\n",
    "            directions += [d]\n",
    "        if colors is None:\n",
    "            colors = []\n",
    "            for i in range(len(exons)):\n",
    "                colors.append(['lightblue', 'lightgreen', 'orange', 'yellow', 'brown'][i % 5])\n",
    "        def get_limits(ex, dir):\n",
    "            start = sys.maxsize\n",
    "            end = -sys.maxsize\n",
    "            for (e, d) in zip(ex, dir):\n",
    "                if d == 1:\n",
    "                    start = min(start,e[0][1])\n",
    "                    end = max(end, e[-1][0])\n",
    "                else:\n",
    "                    start = min(start,e[-1][1])\n",
    "                    end = max(end, e[0][0])\n",
    "            return (start, end)\n",
    "        plt.axes()\n",
    "        plt.xlim((-0.1, 1.1))\n",
    "        plt.ylim((0.1 - 0.5 *  len(exons), 0.3))\n",
    "        plt.margins(0.2)\n",
    "        plt.axis('off')\n",
    "        fig = plt.gcf()\n",
    "        fig.set_size_inches(20, len(exons) * 2)\n",
    "        i = 0\n",
    "        (start, end) = get_limits(exons, directions)\n",
    "        for (ex, di, co, name) in zip(exons, directions, colors, transcripts_id):\n",
    "            self.__draw_exons(ex, di, co, name, offset= -0.5 * i, start_override=start, end_override=end, no_render=True)\n",
    "            i+=1\n",
    "        if _ax is None:\n",
    "            plt.show()\n",
    "        else: \n",
    "            return plt\n",
    "\n",
    "    def draw_transcripts_list(self, gene_name, colors=None):\n",
    "        self.__draw_transcripts_list(gene_name, None, colors)\n",
    "    \n",
    "    def draw_gene_summary(self, gene_name):\n",
    "        ax1 = pw.Brick(figsize=(12,4))\n",
    "        pw.overwrite_axisgrid()\n",
    "        fg = self._trsct_counts_cell_type(gene_name, ax1)\n",
    "        ax4 = pw.load_seaborngrid(fg)\n",
    "        ax1.set_title(\"Transcripts count per cell type\")\n",
    "        ax2 = pw.Brick(figsize=(12,4))\n",
    "        self._plot_transcripts_per_cell_type(gene_name, ax2)\n",
    "        ax2.set_title(\"Transcripts count per cell type\")\n",
    "        ax3 = pw.Brick(figsize=(12,4))\n",
    "        self.__draw_transcripts_list(gene_name, ax3)\n",
    "        ax3.set_title(\"Transcripts list\")\n",
    "        return (ax4/(ax2/ax3)).savefig()\n",
    "    \n",
    "    def __model(self, data, total_counts):\n",
    "        alpha = pyro.param('alpha', torch.ones(data.size(1)), constraint=dist.constraints.positive)\n",
    "        with pyro.plate('data_%i' % (data.size(0)), data.size(0)):\n",
    "            pyro.sample('obs', dist.DirichletMultinomial(concentration=alpha, is_sparse=True, total_count=total_counts), obs=data)\n",
    "    def __guide(self, data, total_counts):\n",
    "        pass\n",
    "    def __perform_mle(self, data, total_counts, num_steps=10000, lr=0.01, tolerance=1e-4, patience=10):\n",
    "        pyro.clear_param_store()\n",
    "        optimizer = Adam({\"lr\": lr})\n",
    "        svi = SVI(self.__model, self.__guide, optimizer, loss=Trace_ELBO())\n",
    "    \n",
    "        best_loss = float('inf')\n",
    "        patience_counter = 0\n",
    "    \n",
    "        for step in range(num_steps):\n",
    "            loss = svi.step(data, total_counts)\n",
    "    \n",
    "            if loss < best_loss - tolerance:\n",
    "                best_loss = loss\n",
    "                patience_counter = 0\n",
    "            else:\n",
    "                patience_counter += 1\n",
    "    \n",
    "            if patience_counter >= patience:\n",
    "                break\n",
    "    \n",
    "        alpha_est = pyro.param('alpha').detach().cpu().numpy()\n",
    "        return best_loss, alpha_est\n",
    "    def __LRT_test(self, data1, data2):\n",
    "        total_counts1 = data1.sum(dim=-1).float()\n",
    "        total_counts2 = data2.sum(dim=-1).float()\n",
    "        combined_data = torch.cat([data1, data2], dim=0)\n",
    "        combined_counts = torch.cat([total_counts1, total_counts2], dim=0)\n",
    "    \n",
    "        loss_full, alpha_full = self.__perform_mle(combined_data, combined_counts)\n",
    "        loss1, alpha1 = self.__perform_mle(data1, total_counts1)\n",
    "        loss2, alpha2 = self.__perform_mle(data2, total_counts2)\n",
    "    \n",
    "        chi2_stat = 2 * (loss_full - (loss1 + loss2))\n",
    "        chi2_stat_tensor = torch.tensor(chi2_stat)\n",
    "    \n",
    "        if chi2_stat_tensor.item() < 0:\n",
    "            return float('nan'), float('nan'), alpha_full, alpha1, alpha2\n",
    "    \n",
    "        p_value = 1 - torch.distributions.Chi2(df=data1.size(1)).cdf(chi2_stat_tensor).item()\n",
    "        \n",
    "        return chi2_stat, p_value, alpha_full, alpha1, alpha2\n",
    "    \n",
    "    def __compare_groups(self, group_1_label, group_2_label, cell_group_column, gene_id):\n",
    "        group_1 = self[self.obs[cell_group_column] == group_1_label]\n",
    "        group_2 = self[self.obs[cell_group_column] == group_2_label]\n",
    "    \n",
    "        data1 = torch.tensor(group_1[:, group_1.var['geneId'] == gene_id].X.toarray(), dtype=torch.float)\n",
    "        data2 = torch.tensor(group_2[:, group_2.var['geneId'] == gene_id].X.toarray(), dtype=torch.float)\n",
    "        \n",
    "        total_counts1 = data1.sum(dim=-1)\n",
    "        total_counts2 = data2.sum(dim=-1)\n",
    "        non_zero_indices1 = total_counts1 > 9\n",
    "        non_zero_indices2 = total_counts2 > 9\n",
    "        data1 = data1[non_zero_indices1]\n",
    "        data2 = data2[non_zero_indices2]\n",
    "        total_counts1 = total_counts1[non_zero_indices1].float()\n",
    "        total_counts2 = total_counts2[non_zero_indices2].float()\n",
    "        \n",
    "    \n",
    "        if data1.size(0) == 0 or data2.size(0) == 0:\n",
    "            return None\n",
    "    \n",
    "        chi2_stat, p_value, alpha_combined, alpha1, alpha2 = self.__LRT_test(data1, data2)\n",
    "        if not np.isnan(chi2_stat):\n",
    "            return {\n",
    "                \"chi2_stat\": chi2_stat,\n",
    "                \"p_value\": p_value,\n",
    "                \"alpha_combined\": alpha_combined.tolist(),\n",
    "                \"alpha1\": alpha1.tolist(),\n",
    "                \"alpha2\": alpha2.tolist()\n",
    "            }\n",
    "        return None\n",
    "    \n",
    "    def __filter_genes(self, group_1_label, group_2_label, cell_group_column):\n",
    "        group_1 = self[self.obs[cell_group_column] == group_1_label]\n",
    "        group_2 = self[self.obs[cell_group_column] == group_2_label]\n",
    "    \n",
    "        gene_counts_group_1 = np.array(group_1.X.sum(axis=0)).flatten()\n",
    "        gene_counts_group_2 = np.array(group_2.X.sum(axis=0)).flatten()\n",
    "        total_gene_counts = gene_counts_group_1 + gene_counts_group_2\n",
    "    \n",
    "        valid_genes = total_gene_counts > 9\n",
    "        adata = self[:, valid_genes]\n",
    "    \n",
    "        group_1 = adata[adata.obs[cell_group_column] == group_1_label]\n",
    "        group_2 = adata[adata.obs[cell_group_column] == group_2_label]\n",
    "    \n",
    "        non_zero_genes_group_1 = np.array((group_1.X != 0).sum(axis=0)).flatten() > 9\n",
    "        non_zero_genes_group_2 = np.array((group_2.X != 0).sum(axis=0)).flatten() > 9\n",
    "        valid_genes = non_zero_genes_group_1 & non_zero_genes_group_2\n",
    "        filtered_with_valid_genes = adata[:, valid_genes]\n",
    "        # print(filtered_with_valid_genes.shape)\n",
    "        for gene_id in np.concatenate((group_1.var['geneId'], group_2.var['geneId'])):\n",
    "            percent_1 = group_1[:, group_1.var['geneId'] == gene_id]\n",
    "            percent_2 = group_2[:, group_2.var['geneId'] == gene_id]\n",
    "            percent_1_X = percent_1.X.astype(float)\n",
    "            percent_2_X = percent_2.X.astype(float)\n",
    "            percent_1_X = percent_1_X / float(sum(percent_1_X.sum(axis=0)))\n",
    "            percent_2_X = percent_2_X / float(sum(percent_2_X.sum(axis=0)))\n",
    "            percent_1.layers['percent'] = percent_1_X\n",
    "            percent_2.layers['percent'] = percent_2_X\n",
    "            diff = False\n",
    "            for transcript_id in np.concatenate((group_1[:, group_1.var['geneId'] == gene_id].var['transcriptId'], group_2[:, group_2.var['geneId'] == gene_id].var['transcriptId']), axis=0):\n",
    "                \n",
    "                per1 = percent_1[:, percent_1.var['transcriptId'] == transcript_id].layers['percent'].sum(axis=0)\n",
    "                per2 = percent_2[:, percent_2.var['transcriptId'] == transcript_id].layers['percent'].sum(axis=0)\n",
    "                diff = abs(per1 - per2) > 0.1\n",
    "                if diff:\n",
    "                    break\n",
    "            if not diff:\n",
    "                filtered_with_valid_genes = filtered_with_valid_genes[:, filtered_with_valid_genes.var['geneId'] != gene_id]\n",
    "            \n",
    "        print(filtered_with_valid_genes.var[['geneId', 'transcriptId']])\n",
    "        return AnnDataIso(filtered_with_valid_genes, self.obs['cell_type'])\n",
    "    \n",
    "    def main(self, cell_group_column):\n",
    "        cell_types = self.obs[cell_group_column].unique()\n",
    "    \n",
    "    \n",
    "        def process_gene(obj, group_1_label, group_2_label, cell_group_column, gene_id):\n",
    "            result = obj.__compare_groups(group_1_label, group_2_label, cell_group_column, gene_id)\n",
    "            if result:\n",
    "                return {\n",
    "                    'gene_id': gene_id,\n",
    "                    'group_1': group_1_label,\n",
    "                    'group_2': group_2_label,\n",
    "                    'p_value': result['p_value'],\n",
    "                    'chi2_stat': result['chi2_stat'],\n",
    "                    'alpha_combined': result['alpha_combined'],\n",
    "                    'alpha1': result['alpha1'],\n",
    "                    'alpha2': result['alpha2'],\n",
    "                    'transcript_ids': obj.var_names[obj.var['geneId'] == gene_id].tolist()\n",
    "                }\n",
    "            return None\n",
    "        \n",
    "        results = []\n",
    "        gene_ids = self.var['geneId'].unique()\n",
    "        print(f\"Processing {len(gene_ids)} genes for {len(cell_types)} cell types...\")\n",
    "        \n",
    "        def add_res(adata, group_1_label, group_2_label, cell_group_column, gene_id, pairwise_results):\n",
    "            pairwise_results.append(process_gene(adata, group_1_label, group_2_label, cell_group_column, gene_id))\n",
    "        \n",
    "        \n",
    "        for group_1_label, group_2_label in combinations(cell_types, 2):\n",
    "            print(f\"Comparing {group_1_label} vs {group_2_label}...\")\n",
    "            filtered_adata = self.__filter_genes(group_1_label, group_2_label, cell_group_column)\n",
    "            pairwise_results = []\n",
    "            \n",
    "            threads = []\n",
    "            for gene_id in gene_ids:\n",
    "                    threads.append(Thread(target=add_res, daemon=True, args=(filtered_adata, group_1_label, group_2_label, cell_group_column, gene_id, pairwise_results)))\n",
    "            for t in threads:\n",
    "                t.start()\n",
    "            for t in threads:\n",
    "                t.join()\n",
    "            \n",
    "            pairwise_results = [res for res in pairwise_results if res is not None]\n",
    "            results.extend(pairwise_results)\n",
    "    \n",
    "        results_df = pd.DataFrame(results)\n",
    "        return results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1601bbb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing 10587 genes for 8 cell types...\n",
      "Comparing Cycling Radial glia vs Radial glia...\n",
      "         geneId           transcriptId\n",
      "1018      Atp5j  ENSMUST00000023608.13\n",
      "1284   Pafah1b3  ENSMUST00000005583.11\n",
      "2888     Atp5g1  ENSMUST00000090541.11\n",
      "3240      Crmp1  ENSMUST00000031004.10\n",
      "6187      Ubxn1   ENSMUST00000096255.5\n",
      "7409        Set   ENSMUST00000067996.6\n",
      "7421    Ngfrap1  ENSMUST00000053540.10\n",
      "7639        Fau   ENSMUST00000178310.7\n",
      "8920        Evl  ENSMUST00000077735.12\n",
      "11932   Smarcb1   ENSMUST00000000925.9\n",
      "13385    Pabpc1  ENSMUST00000001809.14\n",
      "13868   Trappc4   ENSMUST00000034623.7\n",
      "13957       Pkm   ENSMUST00000163694.3\n",
      "15782       Pkm  ENSMUST00000034834.15\n",
      "15928    Zfp428  ENSMUST00000071361.12\n",
      "18453   Anapc11   ENSMUST00000026128.9\n",
      "18784     Rsrp1   ENSMUST00000078084.6\n",
      "20449     Ppdpf  ENSMUST00000016488.12\n",
      "20701    Rbfox2   ENSMUST00000166610.7\n",
      "Comparing Cycling Radial glia vs intermediate progenitor...\n",
      "Empty DataFrame\n",
      "Columns: [geneId, transcriptId]\n",
      "Index: []\n",
      "Comparing Cycling Radial glia vs Imature glutamatergic...\n",
      "        geneId           transcriptId\n",
      "3240     Crmp1  ENSMUST00000031004.10\n",
      "6187     Ubxn1   ENSMUST00000096255.5\n",
      "7383    Rpl35a   ENSMUST00000115075.1\n",
      "7639       Fau   ENSMUST00000178310.7\n",
      "8920       Evl  ENSMUST00000077735.12\n",
      "10788   Rpl35a  ENSMUST00000078804.11\n",
      "10841     Tecr   ENSMUST00000165740.8\n",
      "13385   Pabpc1  ENSMUST00000001809.14\n",
      "13957      Pkm   ENSMUST00000163694.3\n",
      "14001    Hmgb3  ENSMUST00000072699.12\n",
      "15782      Pkm  ENSMUST00000034834.15\n",
      "18453  Anapc11   ENSMUST00000026128.9\n",
      "18784    Rsrp1   ENSMUST00000078084.6\n",
      "20449    Ppdpf  ENSMUST00000016488.12\n",
      "20701   Rbfox2   ENSMUST00000166610.7\n",
      "Comparing Cycling Radial glia vs Mature glutamatergic...\n",
      "        geneId           transcriptId\n",
      "1018     Atp5j  ENSMUST00000023608.13\n",
      "5842      Gnb2   ENSMUST00000150063.8\n",
      "6187     Ubxn1   ENSMUST00000096255.5\n",
      "7321     U2af1   ENSMUST00000014684.4\n",
      "7421   Ngfrap1  ENSMUST00000053540.10\n",
      "10841     Tecr   ENSMUST00000165740.8\n",
      "11179    Srsf3   ENSMUST00000130216.1\n",
      "13385   Pabpc1  ENSMUST00000001809.14\n",
      "13957      Pkm   ENSMUST00000163694.3\n",
      "15782      Pkm  ENSMUST00000034834.15\n",
      "20449    Ppdpf  ENSMUST00000016488.12\n",
      "20701   Rbfox2   ENSMUST00000166610.7\n",
      "Comparing Cycling Radial glia vs mature GABAergic...\n",
      "        geneId           transcriptId\n",
      "1018     Atp5j  ENSMUST00000023608.13\n",
      "2888    Atp5g1  ENSMUST00000090541.11\n",
      "6187     Ubxn1   ENSMUST00000096255.5\n",
      "7321     U2af1   ENSMUST00000014684.4\n",
      "7409       Set   ENSMUST00000067996.6\n",
      "7421   Ngfrap1  ENSMUST00000053540.10\n",
      "7639       Fau   ENSMUST00000178310.7\n",
      "11932  Smarcb1   ENSMUST00000000925.9\n",
      "13385   Pabpc1  ENSMUST00000001809.14\n",
      "15782      Pkm  ENSMUST00000034834.15\n",
      "18236   Atp5c1   ENSMUST00000114897.8\n",
      "18453  Anapc11   ENSMUST00000026128.9\n",
      "Comparing Cycling Radial glia vs Imature GABAergic...\n",
      "        geneId           transcriptId\n",
      "969       Nnat   ENSMUST00000109526.1\n",
      "1018     Atp5j  ENSMUST00000023608.13\n",
      "2888    Atp5g1  ENSMUST00000090541.11\n",
      "6187     Ubxn1   ENSMUST00000096255.5\n",
      "7383    Rpl35a   ENSMUST00000115075.1\n",
      "7409       Set   ENSMUST00000067996.6\n",
      "7421   Ngfrap1  ENSMUST00000053540.10\n",
      "8920       Evl  ENSMUST00000077735.12\n",
      "10788   Rpl35a  ENSMUST00000078804.11\n",
      "10841     Tecr   ENSMUST00000165740.8\n",
      "11932  Smarcb1   ENSMUST00000000925.9\n",
      "13385   Pabpc1  ENSMUST00000001809.14\n",
      "13957      Pkm   ENSMUST00000163694.3\n",
      "15782      Pkm  ENSMUST00000034834.15\n",
      "15928   Zfp428  ENSMUST00000071361.12\n",
      "16000     Nnat   ENSMUST00000153739.8\n",
      "18453  Anapc11   ENSMUST00000026128.9\n",
      "18992   Atp5g2   ENSMUST00000185641.6\n",
      "20449    Ppdpf  ENSMUST00000016488.12\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-55243:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 140, in loss_and_grads\n",
      "    for model_trace, guide_trace in self._get_traces(model, guide, args, kwargs):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/elbo.py\", line 236, in _get_traces\n",
      "    yield self._get_trace(model, guide, args, kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 57, in _get_trace\n",
      "    model_trace, guide_trace = get_importance_trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/enum.py\", line 65, in get_importance_trace\n",
      "    model_trace = poutine.trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 198, in get_trace\n",
      "    self(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 180, in __call__\n",
      "    raise exc from e\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "Trace Shapes:  \n",
      " Param Sites:  \n",
      "        alpha 2\n",
      "Sample Sites:  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing Cycling Radial glia vs Cajal Retzius...\n",
      "        geneId           transcriptId\n",
      "13        Rbm3   ENSMUST00000115616.7\n",
      "1018     Atp5j  ENSMUST00000023608.13\n",
      "1790    Dnaja1   ENSMUST00000030118.9\n",
      "3240     Crmp1  ENSMUST00000031004.10\n",
      "7321     U2af1   ENSMUST00000014684.4\n",
      "7383    Rpl35a   ENSMUST00000115075.1\n",
      "7421   Ngfrap1  ENSMUST00000053540.10\n",
      "10788   Rpl35a  ENSMUST00000078804.11\n",
      "11179    Srsf3   ENSMUST00000130216.1\n",
      "13385   Pabpc1  ENSMUST00000001809.14\n",
      "14420     Meg3   ENSMUST00000143836.7\n",
      "15928   Zfp428  ENSMUST00000071361.12\n",
      "16928     Meg3   ENSMUST00000129245.7\n",
      "17804    Rpl27  ENSMUST00000077856.12\n",
      "18938    Rpl27   ENSMUST00000107249.7\n",
      "20449    Ppdpf  ENSMUST00000016488.12\n",
      "20701   Rbfox2   ENSMUST00000166610.7\n",
      "Comparing Radial glia vs intermediate progenitor...\n",
      "Empty DataFrame\n",
      "Columns: [geneId, transcriptId]\n",
      "Index: []\n",
      "Comparing Radial glia vs Imature glutamatergic...\n",
      "         geneId           transcriptId\n",
      "143       H2afv   ENSMUST00000109737.8\n",
      "385      Mrps21   ENSMUST00000067298.4\n",
      "505        Ufc1   ENSMUST00000111302.3\n",
      "1284   Pafah1b3  ENSMUST00000005583.11\n",
      "1324      Eif4h   ENSMUST00000202622.3\n",
      "...         ...                    ...\n",
      "20254     Rbm4b   ENSMUST00000036744.7\n",
      "20441    Pantr1   ENSMUST00000181725.7\n",
      "20658    Psenen   ENSMUST00000043898.7\n",
      "20701    Rbfox2   ENSMUST00000166610.7\n",
      "20740      Lmo4   ENSMUST00000120539.7\n",
      "\n",
      "[96 rows x 2 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-86236:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "Exception in thread Thread-86439:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing Radial glia vs Mature glutamatergic...\n",
      "       geneId           transcriptId\n",
      "13       Rbm3   ENSMUST00000115616.7\n",
      "385    Mrps21   ENSMUST00000067298.4\n",
      "505      Ufc1   ENSMUST00000111302.3\n",
      "969      Nnat   ENSMUST00000109526.1\n",
      "1324    Eif4h   ENSMUST00000202622.3\n",
      "...       ...                    ...\n",
      "18647   Pnisr  ENSMUST00000029911.11\n",
      "19047    Nfib   ENSMUST00000135024.1\n",
      "19282  Mrpl30  ENSMUST00000027256.11\n",
      "20082   Parp6  ENSMUST00000026267.15\n",
      "20441  Pantr1   ENSMUST00000181725.7\n",
      "\n",
      "[72 rows x 2 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-96306:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "Exception in thread Thread-98836:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing Radial glia vs mature GABAergic...\n",
      "         geneId           transcriptId\n",
      "143       H2afv   ENSMUST00000109737.8\n",
      "1284   Pafah1b3  ENSMUST00000005583.11\n",
      "1324      Eif4h   ENSMUST00000202622.3\n",
      "1394      Dtymk   ENSMUST00000112890.2\n",
      "1858      Mtch1  ENSMUST00000095427.10\n",
      "...         ...                    ...\n",
      "19245    Ube2v2   ENSMUST00000115777.9\n",
      "19446     Apex1  ENSMUST00000049411.11\n",
      "19975      Acp1  ENSMUST00000062740.14\n",
      "20441    Pantr1   ENSMUST00000181725.7\n",
      "20449     Ppdpf  ENSMUST00000016488.12\n",
      "\n",
      "[74 rows x 2 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-107666:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "Exception in thread Thread-107736:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "Exception in thread Thread-108108:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "Exception in thread Thread-108747:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "Exception in thread Thread-112095:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "Exception in thread Thread-111843:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 328, in __model\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/indep_messenger.py\", line 96, in __exit__\n",
      "    return super().__exit__(*args)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 140, in loss_and_grads\n",
      "    for model_trace, guide_trace in self._get_traces(model, guide, args, kwargs):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/elbo.py\", line 236, in _get_traces\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "    yield self._get_trace(model, guide, args, kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 57, in _get_trace\n",
      "    model_trace, guide_trace = get_importance_trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/enum.py\", line 65, in get_importance_trace\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "    model_trace = poutine.trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 198, in get_trace\n",
      "    self(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 180, in __call__\n",
      "    raise exc from e\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 140, in loss_and_grads\n",
      "    for model_trace, guide_trace in self._get_traces(model, guide, args, kwargs):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/elbo.py\", line 236, in _get_traces\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 328, in __model\n",
      "    yield self._get_trace(model, guide, args, kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 57, in _get_trace\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/indep_messenger.py\", line 96, in __exit__\n",
      "    return super().__exit__(*args)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "Trace Shapes:       \n",
      " Param Sites:       \n",
      "        alpha    1  \n",
      "Sample Sites:       \n",
      "    data dist    |  \n",
      "        value 18 |  \n",
      "     obs dist 18 | 1\n",
      "        value 18 | 1\n",
      "    model_trace, guide_trace = get_importance_trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/enum.py\", line 65, in get_importance_trace\n",
      "    model_trace = poutine.trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 198, in get_trace\n",
      "    self(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 180, in __call__\n",
      "    raise exc from e\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "Trace Shapes:     \n",
      " Param Sites:     \n",
      "        alpha    1\n",
      "Sample Sites:     \n",
      "    data dist    |\n",
      "        value 18 |\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing Radial glia vs Imature GABAergic...\n",
      "          geneId           transcriptId\n",
      "114       Hnrnpk   ENSMUST00000176207.7\n",
      "385       Mrps21   ENSMUST00000067298.4\n",
      "969         Nnat   ENSMUST00000109526.1\n",
      "1073   Hnrnpa2b1   ENSMUST00000114459.7\n",
      "1324       Eif4h   ENSMUST00000202622.3\n",
      "...          ...                    ...\n",
      "19820     Nap1l4   ENSMUST00000072727.6\n",
      "19975       Acp1  ENSMUST00000062740.14\n",
      "20157        Bax   ENSMUST00000033093.9\n",
      "20396       Nxf1   ENSMUST00000010248.3\n",
      "20441     Pantr1   ENSMUST00000181725.7\n",
      "\n",
      "[114 rows x 2 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-116667:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "Exception in thread Thread-117576:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 140, in loss_and_grads\n",
      "    for model_trace, guide_trace in self._get_traces(model, guide, args, kwargs):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/elbo.py\", line 236, in _get_traces\n",
      "    yield self._get_trace(model, guide, args, kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 57, in _get_trace\n",
      "    model_trace, guide_trace = get_importance_trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/enum.py\", line 65, in get_importance_trace\n",
      "    model_trace = poutine.trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 198, in get_trace\n",
      "    self(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 180, in __call__\n",
      "    raise exc from e\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "Trace Shapes:  \n",
      " Param Sites:  \n",
      "        alpha 2\n",
      "Sample Sites:  \n",
      "Exception in thread Thread-117480:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "Exception in thread Thread-118253:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-122331:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "Exception in thread Thread-118695:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 328, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/indep_messenger.py\", line 96, in __exit__\n",
      "    return super().__exit__(*args)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 360, in __LRT_test\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 140, in loss_and_grads\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "    for model_trace, guide_trace in self._get_traces(model, guide, args, kwargs):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/elbo.py\", line 236, in _get_traces\n",
      "    yield self._get_trace(model, guide, args, kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 57, in _get_trace\n",
      "    model_trace, guide_trace = get_importance_trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/enum.py\", line 65, in get_importance_trace\n",
      "    model_trace = poutine.trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 198, in get_trace\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 140, in loss_and_grads\n",
      "    self(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 180, in __call__\n",
      "    raise exc from e\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    for model_trace, guide_trace in self._get_traces(model, guide, args, kwargs):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/elbo.py\", line 236, in _get_traces\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    yield self._get_trace(model, guide, args, kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 57, in _get_trace\n",
      "    model_trace, guide_trace = get_importance_trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/enum.py\", line 65, in get_importance_trace\n",
      "    model_trace = poutine.trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 198, in get_trace\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "    self(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 180, in __call__\n",
      "    raise exc from e\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 328, in __model\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/indep_messenger.py\", line 96, in __exit__\n",
      "    return super().__exit__(*args)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "Trace Shapes:      \n",
      " Param Sites:      \n",
      "        alpha   2  \n",
      "Sample Sites:      \n",
      "    data dist   |  \n",
      "        value 1 |  \n",
      "     obs dist 1 | 2\n",
      "        value 1 | 2\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "Trace Shapes:    \n",
      " Param Sites:    \n",
      "        alpha   2\n",
      "Sample Sites:    \n",
      "    data dist   |\n",
      "        value 1 |\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing Radial glia vs Cajal Retzius...\n",
      "         geneId           transcriptId\n",
      "13         Rbm3   ENSMUST00000115616.7\n",
      "267      Cacnb3   ENSMUST00000230490.1\n",
      "385      Mrps21   ENSMUST00000067298.4\n",
      "1284   Pafah1b3  ENSMUST00000005583.11\n",
      "1548      Cdc42   ENSMUST00000030417.9\n",
      "1790     Dnaja1   ENSMUST00000030118.9\n",
      "1858      Mtch1  ENSMUST00000095427.10\n",
      "3154     Polr2i  ENSMUST00000019882.15\n",
      "3582        Ogt   ENSMUST00000155792.1\n",
      "3705     Dnaja1   ENSMUST00000164233.7\n",
      "3873       Lsm6   ENSMUST00000051867.6\n",
      "4539       Ly6h   ENSMUST00000127095.7\n",
      "4614       Fis1   ENSMUST00000019198.6\n",
      "5143      Actr3   ENSMUST00000178474.7\n",
      "5563      Mtch1   ENSMUST00000118366.7\n",
      "5590      Psma3   ENSMUST00000160027.7\n",
      "5786      Ube2b  ENSMUST00000020657.12\n",
      "6184   Pafah1b3   ENSMUST00000148150.7\n",
      "6473      Psmd4   ENSMUST00000107237.7\n",
      "6723     Ppp1cc  ENSMUST00000102528.10\n",
      "7209     Serbp1   ENSMUST00000203077.2\n",
      "7321      U2af1   ENSMUST00000014684.4\n",
      "7639        Fau   ENSMUST00000178310.7\n",
      "8546      Rufy3  ENSMUST00000031229.10\n",
      "8920        Evl  ENSMUST00000077735.12\n",
      "9468     Mpped2   ENSMUST00000111063.7\n",
      "9789     Txndc9   ENSMUST00000162031.7\n",
      "10148      Mbd3   ENSMUST00000105349.7\n",
      "10982    Podxl2   ENSMUST00000145944.2\n",
      "11179     Srsf3   ENSMUST00000130216.1\n",
      "11818     Dgcr6   ENSMUST00000151266.7\n",
      "12734     Arpc2  ENSMUST00000006467.13\n",
      "12930    Hnrnpk   ENSMUST00000224182.1\n",
      "13036      Myl6   ENSMUST00000164181.1\n",
      "13076    Plgrkt   ENSMUST00000143467.7\n",
      "13957       Pkm   ENSMUST00000163694.3\n",
      "14236      Cbx1   ENSMUST00000093943.9\n",
      "15189     Cdc42  ENSMUST00000051477.12\n",
      "15225    Hnrnpk  ENSMUST00000043269.13\n",
      "15243    Serbp1   ENSMUST00000204293.2\n",
      "15415    Rbfox2   ENSMUST00000228087.1\n",
      "15645      Tcf4   ENSMUST00000114977.4\n",
      "15782       Pkm  ENSMUST00000034834.15\n",
      "16129   Prkar1a   ENSMUST00000049527.6\n",
      "16649   Hnrnpdl   ENSMUST00000128187.7\n",
      "17113    Polr2k   ENSMUST00000180159.7\n",
      "17476   Rpl22l1   ENSMUST00000194649.1\n",
      "17715     Arpc2   ENSMUST00000113820.8\n",
      "17971     Bcas2  ENSMUST00000005830.14\n",
      "17995     Ctbp1   ENSMUST00000202868.3\n",
      "18446   Lamtor4   ENSMUST00000062067.7\n",
      "18784     Rsrp1   ENSMUST00000078084.6\n",
      "19245    Ube2v2   ENSMUST00000115777.9\n",
      "19506      Cltb  ENSMUST00000091609.10\n",
      "19634    Mrps24   ENSMUST00000154330.1\n",
      "19975      Acp1  ENSMUST00000062740.14\n",
      "20082     Parp6  ENSMUST00000026267.15\n",
      "20396      Nxf1   ENSMUST00000010248.3\n",
      "20701    Rbfox2   ENSMUST00000166610.7\n",
      "Comparing intermediate progenitor vs Imature glutamatergic...\n",
      "Empty DataFrame\n",
      "Columns: [geneId, transcriptId]\n",
      "Index: []\n",
      "Comparing intermediate progenitor vs Mature glutamatergic...\n",
      "Empty DataFrame\n",
      "Columns: [geneId, transcriptId]\n",
      "Index: []\n",
      "Comparing intermediate progenitor vs mature GABAergic...\n",
      "Empty DataFrame\n",
      "Columns: [geneId, transcriptId]\n",
      "Index: []\n",
      "Comparing intermediate progenitor vs Imature GABAergic...\n",
      "Empty DataFrame\n",
      "Columns: [geneId, transcriptId]\n",
      "Index: []\n",
      "Comparing intermediate progenitor vs Cajal Retzius...\n",
      "Empty DataFrame\n",
      "Columns: [geneId, transcriptId]\n",
      "Index: []\n",
      "Comparing Imature glutamatergic vs Mature glutamatergic...\n",
      "        geneId           transcriptId\n",
      "505       Ufc1   ENSMUST00000111302.3\n",
      "2663      Myl6   ENSMUST00000218127.1\n",
      "3154    Polr2i  ENSMUST00000019882.15\n",
      "3582       Ogt   ENSMUST00000155792.1\n",
      "4614      Fis1   ENSMUST00000019198.6\n",
      "5590     Psma3   ENSMUST00000160027.7\n",
      "5786     Ube2b  ENSMUST00000020657.12\n",
      "6473     Psmd4   ENSMUST00000107237.7\n",
      "7321     U2af1   ENSMUST00000014684.4\n",
      "7800   Hnrnph3  ENSMUST00000020263.13\n",
      "7816      Cuta   ENSMUST00000025027.8\n",
      "7912     Pdcd6   ENSMUST00000022060.6\n",
      "7933    Polr1d   ENSMUST00000110557.4\n",
      "8470   Krtcap2  ENSMUST00000040888.11\n",
      "10148     Mbd3   ENSMUST00000105349.7\n",
      "10662    Rab3a  ENSMUST00000034301.11\n",
      "11179    Srsf3   ENSMUST00000130216.1\n",
      "11502    Erp29   ENSMUST00000130451.1\n",
      "11764     Hes6   ENSMUST00000086851.1\n",
      "11818    Dgcr6   ENSMUST00000151266.7\n",
      "11833    Gpm6b   ENSMUST00000112228.7\n",
      "12930   Hnrnpk   ENSMUST00000224182.1\n",
      "13036     Myl6   ENSMUST00000164181.1\n",
      "13076   Plgrkt   ENSMUST00000143467.7\n",
      "13097   Polr1d   ENSMUST00000050970.3\n",
      "13125    Ube2i   ENSMUST00000174031.7\n",
      "13282   Anapc5   ENSMUST00000086216.8\n",
      "13332     Tcf4   ENSMUST00000114978.8\n",
      "13957      Pkm   ENSMUST00000163694.3\n",
      "14264     Mtdh   ENSMUST00000172082.1\n",
      "14777     Lsm6   ENSMUST00000130325.7\n",
      "15074    Hspa5  ENSMUST00000028222.12\n",
      "15225   Hnrnpk  ENSMUST00000043269.13\n",
      "15415   Rbfox2   ENSMUST00000228087.1\n",
      "15645     Tcf4   ENSMUST00000114977.4\n",
      "15782      Pkm  ENSMUST00000034834.15\n",
      "16263   Rbmxl1   ENSMUST00000211719.1\n",
      "16656    Rbm39   ENSMUST00000128243.1\n",
      "16704  Hdgfrp3   ENSMUST00000149288.1\n",
      "18453  Anapc11   ENSMUST00000026128.9\n",
      "19047     Nfib   ENSMUST00000135024.1\n",
      "20254    Rbm4b   ENSMUST00000036744.7\n",
      "20701   Rbfox2   ENSMUST00000166610.7\n",
      "Comparing Imature glutamatergic vs mature GABAergic...\n",
      "        geneId           transcriptId\n",
      "2523     Gsk3b  ENSMUST00000023507.12\n",
      "2663      Myl6   ENSMUST00000218127.1\n",
      "4614      Fis1   ENSMUST00000019198.6\n",
      "5951    Rpl35a   ENSMUST00000115079.7\n",
      "6167      Tecr  ENSMUST00000019382.16\n",
      "7321     U2af1   ENSMUST00000014684.4\n",
      "7383    Rpl35a   ENSMUST00000115075.1\n",
      "7409       Set   ENSMUST00000067996.6\n",
      "7890     Emc10   ENSMUST00000118808.7\n",
      "7912     Pdcd6   ENSMUST00000022060.6\n",
      "8470   Krtcap2  ENSMUST00000040888.11\n",
      "9468    Mpped2   ENSMUST00000111063.7\n",
      "10149   Pea15a  ENSMUST00000013842.11\n",
      "10788   Rpl35a  ENSMUST00000078804.11\n",
      "11502    Erp29   ENSMUST00000130451.1\n",
      "11764     Hes6   ENSMUST00000086851.1\n",
      "11818    Dgcr6   ENSMUST00000151266.7\n",
      "11932  Smarcb1   ENSMUST00000000925.9\n",
      "11986     Gnas   ENSMUST00000109087.7\n",
      "12930   Hnrnpk   ENSMUST00000224182.1\n",
      "13036     Myl6   ENSMUST00000164181.1\n",
      "13332     Tcf4   ENSMUST00000114978.8\n",
      "13486     Mpst   ENSMUST00000229739.1\n",
      "13752    Vps29   ENSMUST00000155671.7\n",
      "14001    Hmgb3  ENSMUST00000072699.12\n",
      "14111   Atp5g3   ENSMUST00000018914.2\n",
      "15189    Cdc42  ENSMUST00000051477.12\n",
      "15225   Hnrnpk  ENSMUST00000043269.13\n",
      "15639     Clta   ENSMUST00000170241.7\n",
      "15782      Pkm  ENSMUST00000034834.15\n",
      "18236   Atp5c1   ENSMUST00000114897.8\n",
      "18453  Anapc11   ENSMUST00000026128.9\n",
      "19282   Mrpl30  ENSMUST00000027256.11\n",
      "20449    Ppdpf  ENSMUST00000016488.12\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-201363:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n",
      "Exception in thread Thread-203461:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 140, in loss_and_grads\n",
      "    for model_trace, guide_trace in self._get_traces(model, guide, args, kwargs):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/elbo.py\", line 236, in _get_traces\n",
      "    yield self._get_trace(model, guide, args, kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 57, in _get_trace\n",
      "    model_trace, guide_trace = get_importance_trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/enum.py\", line 65, in get_importance_trace\n",
      "    model_trace = poutine.trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 198, in get_trace\n",
      "    self(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 180, in __call__\n",
      "    raise exc from e\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "Trace Shapes:  \n",
      " Param Sites:  \n",
      "        alpha 3\n",
      "Sample Sites:  \n",
      "Exception in thread Thread-203019:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 230, in compute_log_prob\n",
      "    log_p = site[\"fn\"].log_prob(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/distributions/conjugate.py\", line 208, in log_prob\n",
      "    self._validate_sample(value)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/torch/distributions/distribution.py\", line 281, in _validate_sample\n",
      "    raise ValueError('The right-most size of value must match event_shape: {} vs {}.'.\n",
      "ValueError: The right-most size of value must match event_shape: torch.Size([4, 1]) vs torch.Size([3]).\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 140, in loss_and_grads\n",
      "    for model_trace, guide_trace in self._get_traces(model, guide, args, kwargs):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/elbo.py\", line 236, in _get_traces\n",
      "    yield self._get_trace(model, guide, args, kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 57, in _get_trace\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "    model_trace, guide_trace = get_importance_trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/enum.py\", line 75, in get_importance_trace\n",
      "    model_trace.compute_log_prob()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 236, in compute_log_prob\n",
      "    raise ValueError(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 230, in compute_log_prob\n",
      "    log_p = site[\"fn\"].log_prob(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/distributions/conjugate.py\", line 208, in log_prob\n",
      "    self._validate_sample(value)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/torch/distributions/distribution.py\", line 281, in _validate_sample\n",
      "    raise ValueError('The right-most size of value must match event_shape: {} vs {}.'.\n",
      "ValueError: Error while computing log_prob at site 'obs':\n",
      "The right-most size of value must match event_shape: torch.Size([4, 1]) vs torch.Size([3]).\n",
      "Trace Shapes:      \n",
      " Param Sites:      \n",
      "        alpha   3  \n",
      "Sample Sites:      \n",
      "     obs dist 4 | 3\n",
      "        value 4 | 1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Comparing Imature glutamatergic vs Imature GABAergic...\n",
      "        geneId           transcriptId\n",
      "114     Hnrnpk   ENSMUST00000176207.7\n",
      "969       Nnat   ENSMUST00000109526.1\n",
      "1198      Lsm7   ENSMUST00000035775.8\n",
      "1517    Zfp428   ENSMUST00000177205.1\n",
      "1548     Cdc42   ENSMUST00000030417.9\n",
      "1600      Nasp   ENSMUST00000154811.7\n",
      "1685    Mrps12   ENSMUST00000056078.8\n",
      "1819      Nasp  ENSMUST00000030457.11\n",
      "1858     Mtch1  ENSMUST00000095427.10\n",
      "2527     Rps28   ENSMUST00000173844.7\n",
      "2663      Myl6   ENSMUST00000218127.1\n",
      "3139      Cfl2   ENSMUST00000078124.7\n",
      "3240     Crmp1  ENSMUST00000031004.10\n",
      "3723     Mat2a   ENSMUST00000059472.9\n",
      "3911      Rer1   ENSMUST00000030914.3\n",
      "4324      Rtn4   ENSMUST00000102843.9\n",
      "4567    Ube2d3   ENSMUST00000197859.4\n",
      "5563     Mtch1   ENSMUST00000118366.7\n",
      "6167      Tecr  ENSMUST00000019382.16\n",
      "6770    Mrpl51   ENSMUST00000032485.6\n",
      "7639       Fau   ENSMUST00000178310.7\n",
      "7912     Pdcd6   ENSMUST00000022060.6\n",
      "8039     Tra2b   ENSMUST00000161286.7\n",
      "8353    Hnrnpc   ENSMUST00000227242.1\n",
      "8920       Evl  ENSMUST00000077735.12\n",
      "8930    Atp5g2   ENSMUST00000075630.9\n",
      "9208     Rab6a   ENSMUST00000032946.9\n",
      "9468    Mpped2   ENSMUST00000111063.7\n",
      "10537    Acin1   ENSMUST00000111484.8\n",
      "10841     Tecr   ENSMUST00000165740.8\n",
      "11502    Erp29   ENSMUST00000130451.1\n",
      "11833    Gpm6b   ENSMUST00000112228.7\n",
      "11932  Smarcb1   ENSMUST00000000925.9\n",
      "12185    Psmg4   ENSMUST00000147632.2\n",
      "12544      Tdg   ENSMUST00000151390.7\n",
      "12930   Hnrnpk   ENSMUST00000224182.1\n",
      "13036     Myl6   ENSMUST00000164181.1\n",
      "13332     Tcf4   ENSMUST00000114978.8\n",
      "13957      Pkm   ENSMUST00000163694.3\n",
      "14472   Mrpl21   ENSMUST00000139844.7\n",
      "15189    Cdc42  ENSMUST00000051477.12\n",
      "15225   Hnrnpk  ENSMUST00000043269.13\n",
      "15639     Clta   ENSMUST00000170241.7\n",
      "15645     Tcf4   ENSMUST00000114977.4\n",
      "15782      Pkm  ENSMUST00000034834.15\n",
      "15868     Clta   ENSMUST00000107849.9\n",
      "15909   Mrpl15   ENSMUST00000156816.6\n",
      "15928   Zfp428  ENSMUST00000071361.12\n",
      "16000     Nnat   ENSMUST00000153739.8\n",
      "16263   Rbmxl1   ENSMUST00000211719.1\n",
      "16656    Rbm39   ENSMUST00000128243.1\n",
      "18153    l7Rn6   ENSMUST00000153470.8\n",
      "18992   Atp5g2   ENSMUST00000185641.6\n",
      "19282   Mrpl30  ENSMUST00000027256.11\n",
      "19648    Acbd6  ENSMUST00000080138.12\n",
      "19820   Nap1l4   ENSMUST00000072727.6\n",
      "20254    Rbm4b   ENSMUST00000036744.7\n",
      "20396     Nxf1   ENSMUST00000010248.3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Exception in thread Thread-211950:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 140, in loss_and_grads\n",
      "    for model_trace, guide_trace in self._get_traces(model, guide, args, kwargs):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/elbo.py\", line 236, in _get_traces\n",
      "    yield self._get_trace(model, guide, args, kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/trace_elbo.py\", line 57, in _get_trace\n",
      "    model_trace, guide_trace = get_importance_trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/enum.py\", line 65, in get_importance_trace\n",
      "    model_trace = poutine.trace(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 198, in get_trace\n",
      "    self(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 180, in __call__\n",
      "    raise exc from e\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 174, in __call__\n",
      "    ret = self.fn(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 12, in _context_wrap\n",
      "    return fn(*args, **kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 327, in __model\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 88, in __init__\n",
      "    self.size, self.subsample_size, self._indices = self._subsample(\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/subsample_messenger.py\", line 126, in _subsample\n",
      "    apply_stack(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/runtime.py\", line 221, in apply_stack\n",
      "    frame._postprocess_message(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 168, in _postprocess_message\n",
      "    return method(msg)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 135, in _pyro_post_sample\n",
      "    self.trace.add_node(msg[\"name\"], **msg.copy())\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_struct.py\", line 115, in add_node\n",
      "    raise RuntimeError(\n",
      "RuntimeError: Multiple sample sites named 'data'\n",
      "Trace Shapes:  \n",
      " Param Sites:  \n",
      "        alpha 2\n",
      "Sample Sites:  \n",
      "Exception in thread Thread-212763:\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 980, in _bootstrap_inner\n",
      "    self.run()\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/threading.py\", line 917, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 471, in add_res\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 451, in process_gene\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 393, in __compare_groups\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 359, in __LRT_test\n",
      "  File \"/tmp/ipykernel_511566/2694172863.py\", line 340, in __perform_mle\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/infer/svi.py\", line 145, in step\n",
      "    loss = self.loss_and_grads(self.model, self.guide, *args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/trace_messenger.py\", line 97, in __exit__\n",
      "    return super().__exit__(*args, **kwargs)\n",
      "  File \"/home/diamant/.conda/envs/iso_swt/lib/python3.9/site-packages/pyro/poutine/messenger.py\", line 138, in __exit__\n",
      "    raise ValueError(\"This Messenger is not on the bottom of the stack\")\n",
      "ValueError: This Messenger is not on the bottom of the stack\n"
     ]
    }
   ],
   "source": [
    "adata_m = AnnDataIso(adata_iso, cell_type['leiden'].tolist())\n",
    "results = adata_m.main(\"cell_type\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df958ee5",
   "metadata": {},
   "outputs": [],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ac4ec19",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
