{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "description: This module focuses on loading data into **Allos**, providing a suite\n",
    "  of reader functions to accommodate a variety of formats for isoform-resolution single-cell\n",
    "  data. It is designed to streamline the import process from multiple protocols that\n",
    "  generate such data and offers tools to download test and example datasets for ease\n",
    "  of experimentation.\n",
    "output-file: readers_tests.html\n",
    "title: Readers\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "has_sd": true,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/readers_tests.py#L26){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_resource_path\n",
       "\n",
       ">      get_resource_path (filename)\n",
       "\n",
       "*Dynamically find the correct path to the 'resources/' directory based on execution location.*"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/readers_tests.py#L26){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### get_resource_path\n",
       "\n",
       ">      get_resource_path (filename)\n",
       "\n",
       "*Dynamically find the correct path to the 'resources/' directory based on execution location.*"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(get_resource_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "949e6d31",
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "ðŸ”Ž Looking for file at: /data/analysis/data_mcandrew/allos_2/allos/resources/e18.mouse.clusters.csv\n",
      "âœ… File found at: /data/analysis/data_mcandrew/allos_2/allos/resources/e18.mouse.clusters.csv\n"
     ]
    }
   ],
   "source": [
    "file_path = get_resource_path(\"e18.mouse.clusters.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "has_sd": true,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/readers_tests.py#L69){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### download_test_data\n",
       "\n",
       ">      download_test_data (url:str='https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM\n",
       ">                          3748nnn/GSM3748087/suppl/GSM3748087%5F190c.isoforms.m\n",
       ">                          atrix.txt.gz', output_filename:str=None,\n",
       ">                          decompress:bool=True)\n",
       "\n",
       "*Download test data to the correct directory, dynamically adjusting based on the execution context.\n",
       "Optionally decompresses gzip files if detected.*\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| url | str | https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3748nnn/GSM3748087/suppl/GSM3748087%5F190c.isoforms.matrix.txt.gz | URL to download the data from. |\n",
       "| output_filename | str | None | Name of the file to save the data as (default: name from the URL). |\n",
       "| decompress | bool | True | Whether to decompress the file if it is a gzip archive (default True). |\n",
       "| **Returns** | **str** |  | **Path to the downloaded or decompressed file.** |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/readers_tests.py#L69){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### download_test_data\n",
       "\n",
       ">      download_test_data (url:str='https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM\n",
       ">                          3748nnn/GSM3748087/suppl/GSM3748087%5F190c.isoforms.m\n",
       ">                          atrix.txt.gz', output_filename:str=None,\n",
       ">                          decompress:bool=True)\n",
       "\n",
       "*Download test data to the correct directory, dynamically adjusting based on the execution context.\n",
       "Optionally decompresses gzip files if detected.*\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| url | str | https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3748nnn/GSM3748087/suppl/GSM3748087%5F190c.isoforms.matrix.txt.gz | URL to download the data from. |\n",
       "| output_filename | str | None | Name of the file to save the data as (default: name from the URL). |\n",
       "| decompress | bool | True | Whether to decompress the file if it is a gzip archive (default True). |\n",
       "| **Returns** | **str** |  | **Path to the downloaded or decompressed file.** |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(download_test_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "has_sd": true,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/readers_tests.py#L141){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### iso_concat\n",
       "\n",
       ">      iso_concat (data_inputs, batch_info=None, batch_type='path')\n",
       "\n",
       "*Concatenates a list of AnnData objects or paths to AnnData objects based on the union of transcriptIds,\n",
       "while preserving geneId information which might be non-unique per transcriptId.\n",
       "Missing values are filled with zeros. Adds a batch column to `.obs` based on the file path, obs_names, or numeric.\n",
       "\n",
       "Parameters:\n",
       "data_inputs (list of str or AnnData):\n",
       "    List of paths to AnnData objects or AnnData objects to concatenate.\n",
       "batch_info (list of str, optional):\n",
       "    List of batch identifiers for each AnnData object in data_inputs.\n",
       "    If not provided, batch identifiers are extracted from file paths, obs_names, or a numeric sequence.\n",
       "batch_type (str, optional):\n",
       "    Specifies which type of batch information to use. One of ['path', 'obs_names', 'numeric'].\n",
       "    Defaults to 'path'.\n",
       "\n",
       "Returns:\n",
       "AnnData:\n",
       "    A single concatenated AnnData object with harmonized features, geneId annotations, and batch info.*"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/readers_tests.py#L141){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### iso_concat\n",
       "\n",
       ">      iso_concat (data_inputs, batch_info=None, batch_type='path')\n",
       "\n",
       "*Concatenates a list of AnnData objects or paths to AnnData objects based on the union of transcriptIds,\n",
       "while preserving geneId information which might be non-unique per transcriptId.\n",
       "Missing values are filled with zeros. Adds a batch column to `.obs` based on the file path, obs_names, or numeric.\n",
       "\n",
       "Parameters:\n",
       "data_inputs (list of str or AnnData):\n",
       "    List of paths to AnnData objects or AnnData objects to concatenate.\n",
       "batch_info (list of str, optional):\n",
       "    List of batch identifiers for each AnnData object in data_inputs.\n",
       "    If not provided, batch identifiers are extracted from file paths, obs_names, or a numeric sequence.\n",
       "batch_type (str, optional):\n",
       "    Specifies which type of batch information to use. One of ['path', 'obs_names', 'numeric'].\n",
       "    Defaults to 'path'.\n",
       "\n",
       "Returns:\n",
       "AnnData:\n",
       "    A single concatenated AnnData object with harmonized features, geneId annotations, and batch info.*"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(iso_concat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "has_sd": true,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/readers_tests.py#L231){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### read_sicelore_isomatrix\n",
       "\n",
       ">      read_sicelore_isomatrix (file_path:str, gene_id_label:str='geneId',\n",
       ">                               transcript_id_label:str='transcriptId',\n",
       ">                               remove_undef:bool=True, sparse:bool=False)\n",
       "\n",
       "*Read a SiCeLoRe isomatrix file (tab-delimited) and convert it into a scanpy-compatible AnnData object.*\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| file_path | str |  | Path to the isomatrix file (tab-delimited). |\n",
       "| gene_id_label | str | geneId | Row/column label used for gene IDs (default \"geneId\"). |\n",
       "| transcript_id_label | str | transcriptId | Row/column label used for transcript IDs (default \"transcriptId\"). |\n",
       "| remove_undef | bool | True | Whether to remove rows with transcriptId=\"undef\" (default True). |\n",
       "| sparse | bool | False | Whether to store the matrix in sparse format (default False). |\n",
       "| **Returns** | **AnnData** |  | **An AnnData object containing numeric data in `.X` and metadata in `.var`.** |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/readers_tests.py#L231){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### read_sicelore_isomatrix\n",
       "\n",
       ">      read_sicelore_isomatrix (file_path:str, gene_id_label:str='geneId',\n",
       ">                               transcript_id_label:str='transcriptId',\n",
       ">                               remove_undef:bool=True, sparse:bool=False)\n",
       "\n",
       "*Read a SiCeLoRe isomatrix file (tab-delimited) and convert it into a scanpy-compatible AnnData object.*\n",
       "\n",
       "|    | **Type** | **Default** | **Details** |\n",
       "| -- | -------- | ----------- | ----------- |\n",
       "| file_path | str |  | Path to the isomatrix file (tab-delimited). |\n",
       "| gene_id_label | str | geneId | Row/column label used for gene IDs (default \"geneId\"). |\n",
       "| transcript_id_label | str | transcriptId | Row/column label used for transcript IDs (default \"transcriptId\"). |\n",
       "| remove_undef | bool | True | Whether to remove rows with transcriptId=\"undef\" (default True). |\n",
       "| sparse | bool | False | Whether to store the matrix in sparse format (default False). |\n",
       "| **Returns** | **AnnData** |  | **An AnnData object containing numeric data in `.X` and metadata in `.var`.** |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(read_sicelore_isomatrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "has_sd": true,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/readers_tests.py#L318){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### process_mouse_data\n",
       "\n",
       ">      process_mouse_data ()\n",
       "\n",
       "*Downloads test data, reads two mouse isoform count matrices, and merges them into a single AnnData object.\n",
       "It also reads a CSV file containing barcode-to-cell_type mappings, merges this information into the AnnData\n",
       "object's obs DataFrame, and filters out entries with no cell_type assigned.\n",
       "\n",
       "Returns:\n",
       "    combined_mouse_data (AnnData): The merged and annotated AnnData object.*"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/readers_tests.py#L318){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### process_mouse_data\n",
       "\n",
       ">      process_mouse_data ()\n",
       "\n",
       "*Downloads test data, reads two mouse isoform count matrices, and merges them into a single AnnData object.\n",
       "It also reads a CSV file containing barcode-to-cell_type mappings, merges this information into the AnnData\n",
       "object's obs DataFrame, and filters out entries with no cell_type assigned.\n",
       "\n",
       "Returns:\n",
       "    combined_mouse_data (AnnData): The merged and annotated AnnData object.*"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(process_mouse_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9a7184d3",
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "ðŸ”Ž Looking for file at: /data/analysis/data_mcandrew/allos_2/allos/resources/e18.mouse.clusters.csv\n",
      "âœ… File found at: /data/analysis/data_mcandrew/allos_2/allos/resources/e18.mouse.clusters.csv\n",
      "âœ… File already exists at: /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_1.txt.gz\n",
      "\n",
      "ðŸ”„ Decompressing /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_1.txt.gz to /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_1.txt...\n",
      "âœ… Decompression complete.\n",
      "Test data (mouse_1) downloaded successfully\n",
      "âœ… File already exists at: /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_2.txt.gz\n",
      "\n",
      "ðŸ”„ Decompressing /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_2.txt.gz to /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_2.txt...\n",
      "âœ… Decompression complete.\n",
      "Test data (mouse_2) downloaded successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "mouse_data = process_mouse_data()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9249c6250c040eb3",
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "ðŸ”Ž Looking for file at: /data/analysis/data_mcandrew/allos_2/allos/resources/e18.mouse.clusters.csv\n",
      "âœ… File found at: /data/analysis/data_mcandrew/allos_2/allos/resources/e18.mouse.clusters.csv\n",
      "\n",
      "ðŸ”„ Downloading https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3748nnn/GSM3748087/suppl/GSM3748087%5F190c.isoforms.matrix.txt.gz to /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_1.txt.gz...\n",
      "âœ… Download complete.\n",
      "\n",
      "ðŸ”„ Decompressing /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_1.txt.gz to /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_1.txt...\n",
      "âœ… Decompression complete.\n",
      "Test data downloaded successfully\n",
      "\n",
      "ðŸ”„ Downloading https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3748nnn/GSM3748089/suppl/GSM3748089%5F951c.isoforms.matrix.txt.gz to /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_2.txt.gz...\n",
      "âœ… Download complete.\n",
      "\n",
      "ðŸ”„ Decompressing /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_2.txt.gz to /data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_2.txt...\n",
      "âœ… Decompression complete.\n",
      "Test data downloaded successfully\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/analysis/data_mcandrew/Allos_new/allos_env/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "file_path = get_resource_path(\"e18.mouse.clusters.csv\")\n",
    "df = pd.read_csv(file_path)\n",
    "df['barcode'] = df.index.str.split('_').str[1]\n",
    "\n",
    "\n",
    "mouse_data_str_1 = download_test_data(output_filename='mouse_1.txt.gz')\n",
    "print(\"Test data downloaded successfully\")\n",
    "\n",
    "\n",
    "mouse_data_str_2 = download_test_data(\"https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3748nnn/GSM3748089/suppl/GSM3748089%5F951c.isoforms.matrix.txt.gz\", output_filename='mouse_2.txt.gz')\n",
    "print(\"Test data downloaded successfully\")\n",
    "\n",
    "\n",
    "mouse_1 = read_sicelore_isomatrix(file_path=mouse_data_str_1)\n",
    "mouse_2 = read_sicelore_isomatrix(file_path=mouse_data_str_2)\n",
    "\n",
    "combined_mouse_data = iso_concat([mouse_1, mouse_2], batch_type='numeric')\n",
    "\n",
    "\n",
    "combined_mouse_data.obs_names_make_unique()\n",
    "# Step 1: Remove any duplicate barcodes in the DataFrame\n",
    "df_unique = df.drop_duplicates(subset='barcode')\n",
    "\n",
    "# Step 2: Filter the DataFrame to include only the barcodes present in the AnnData object\n",
    "df_filtered = df_unique[df_unique['barcode'].isin(combined_mouse_data.obs_names)]\n",
    "\n",
    "# Step 3: Set the index of the filtered DataFrame to 'barcode' to make the merge easier\n",
    "df_filtered.set_index('barcode', inplace=True)\n",
    "\n",
    "# Step 4: Create a DataFrame from the obs DataFrame of the AnnData object to ensure the same index\n",
    "obs_df = combined_mouse_data.obs.copy()\n",
    "\n",
    "# Step 5: Initialize a new column 'cell_type' with NaN values in the obs DataFrame\n",
    "obs_df['cell_type'] = pd.NA\n",
    "\n",
    "# Step 6: Update the 'cell_type' column with values from the filtered DataFrame where indices match\n",
    "obs_df.update(df_filtered['illumina.ident'].rename('cell_type'))\n",
    "\n",
    "# Step 7: Ensure the index is unique and assign the updated DataFrame back to the obs attribute of the AnnData object\n",
    "if obs_df.index.is_unique:\n",
    "    combined_mouse_data.obs = obs_df\n",
    "else:\n",
    "    raise ValueError(\"The index of the obs DataFrame is not unique.\")\n",
    "\n",
    "# Now, the 'cell_type' column should be added to the obs DataFrame of your AnnData object\n",
    "combined_mouse_data = combined_mouse_data[~combined_mouse_data.obs['cell_type'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "57231583",
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/analysis/data_mcandrew/allos_2/allos/nbs/data/mouse_1.txt'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mouse_data_str_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "a2383ac7",
   "metadata": {
    "language": "python"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs Ã— n_vars = 1109 Ã— 31986\n",
       "    obs: 'batch', 'cell_type'\n",
       "    var: 'geneId'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_mouse_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9dfd332",
   "metadata": {
    "language": "python"
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
