{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "has_sd": true,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/switch_search.py#L36){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### SwitchSearch\n",
       "\n",
       ">      SwitchSearch (anndata:anndata._core.anndata.AnnData,\n",
       ">                    cell_types:pandas.core.frame.DataFrame, palette='ghibli')\n",
       "\n",
       "*An annotated data matrix.\n",
       "\n",
       ".. figure:: ../_static/img/anndata_schema.svg\n",
       "   :width: 260px\n",
       "   :align: right\n",
       "   :class: dark-light\n",
       "\n",
       ":class:`~anndata.AnnData` stores a data matrix :attr:`X` together with annotations\n",
       "of observations :attr:`obs` (:attr:`obsm`, :attr:`obsp`),\n",
       "variables :attr:`var` (:attr:`varm`, :attr:`varp`),\n",
       "and unstructured annotations :attr:`uns`.\n",
       "\n",
       "An :class:`~anndata.AnnData` object `adata` can be sliced like a\n",
       ":class:`~pandas.DataFrame`,\n",
       "for instance `adata_subset = adata[:, list_of_variable_names]`.\n",
       ":class:`~anndata.AnnData`’s basic structure is similar to R’s ExpressionSet\n",
       "[Huber15]_. If setting an `.h5ad`-formatted HDF5 backing file `.filename`,\n",
       "data remains on the disk but is automatically loaded into memory if needed.*"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/switch_search.py#L36){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### SwitchSearch\n",
       "\n",
       ">      SwitchSearch (anndata:anndata._core.anndata.AnnData,\n",
       ">                    cell_types:pandas.core.frame.DataFrame, palette='ghibli')\n",
       "\n",
       "*An annotated data matrix.\n",
       "\n",
       ".. figure:: ../_static/img/anndata_schema.svg\n",
       "   :width: 260px\n",
       "   :align: right\n",
       "   :class: dark-light\n",
       "\n",
       ":class:`~anndata.AnnData` stores a data matrix :attr:`X` together with annotations\n",
       "of observations :attr:`obs` (:attr:`obsm`, :attr:`obsp`),\n",
       "variables :attr:`var` (:attr:`varm`, :attr:`varp`),\n",
       "and unstructured annotations :attr:`uns`.\n",
       "\n",
       "An :class:`~anndata.AnnData` object `adata` can be sliced like a\n",
       ":class:`~pandas.DataFrame`,\n",
       "for instance `adata_subset = adata[:, list_of_variable_names]`.\n",
       ":class:`~anndata.AnnData`’s basic structure is similar to R’s ExpressionSet\n",
       "[Huber15]_. If setting an `.h5ad`-formatted HDF5 backing file `.filename`,\n",
       "data remains on the disk but is automatically loaded into memory if needed.*"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(SwitchSearch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 0,
   "has_sd": true,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/switch_search.py#L720){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### create_iso_adata\n",
       "\n",
       ">      create_iso_adata (path:str)\n",
       "\n",
       "*Creates an AnnData object containing isoform counts from a Sicelore output file.\n",
       "\n",
       "Parameters:\n",
       "- path (str): Path to the input file, which is expected to be tab-separated and contain:\n",
       "    - `transcriptId`: Transcript identifiers.\n",
       "    - `geneId`: Gene identifiers.\n",
       "    - Transcript counts for individual barcodes (cells) as additional columns.\n",
       "\n",
       "Returns:\n",
       "- adata_iso (AnnData): An AnnData object with isoform counts.*\n",
       "\n",
       "|    | **Type** | **Details** |\n",
       "| -- | -------- | ----------- |\n",
       "| path | str | Path to a tab-separated file, typically the Sicelore output with transcript counts. |"
      ],
      "text/plain": [
       "---\n",
       "\n",
       "[source](https://github.com/cobioda/allos/blob/main/allos/switch_search.py#L720){target=\"_blank\" style=\"float:right; font-size:smaller\"}\n",
       "\n",
       "### create_iso_adata\n",
       "\n",
       ">      create_iso_adata (path:str)\n",
       "\n",
       "*Creates an AnnData object containing isoform counts from a Sicelore output file.\n",
       "\n",
       "Parameters:\n",
       "- path (str): Path to the input file, which is expected to be tab-separated and contain:\n",
       "    - `transcriptId`: Transcript identifiers.\n",
       "    - `geneId`: Gene identifiers.\n",
       "    - Transcript counts for individual barcodes (cells) as additional columns.\n",
       "\n",
       "Returns:\n",
       "- adata_iso (AnnData): An AnnData object with isoform counts.*\n",
       "\n",
       "|    | **Type** | **Details** |\n",
       "| -- | -------- | ----------- |\n",
       "| path | str | Path to a tab-separated file, typically the Sicelore output with transcript counts. |"
      ]
     },
     "execution_count": null,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#| echo: false\n",
    "#| output: asis\n",
    "show_doc(create_iso_adata)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
