{
 "cells": [
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "description: This module focuses on the **GeneReports** class, which extends the capabilities\n",
    "  of the AnnDataIso and TranscriptPlots classes to provide comprehensive gene-level\n",
    "  analysis and visualization tools. It is specifically designed to handle gene expression\n",
    "  data, offering specialized methods to efficiently manage and analyze data from various\n",
    "  single-cell RNA sequencing protocols, facilitating seamless integration and manipulation\n",
    "  within the Allos framework.\n",
    "output-file: gene_report.html\n",
    "title: Gene Report Module\n",
    "\n",
    "---\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d463bb2dca533986",
   "metadata": {
    "language": "python"
   },
   "outputs": [],
   "source": [
    "class GeneReports(TranscriptPlots, AnnDataIso):\n",
    "    def __init__(self, anndata: ad.AnnData, cell_types: pd.DataFrame, palette='ghibli', gtf_file=None, reference_fasta=None):\n",
    "        AnnDataIso.__init__(self,anndata, cell_types, palette)\n",
    "        TranscriptPlots.__init__(self,gtf_file, reference_fasta)\n",
    "\n",
    "    # Plotting a summary of isoform-related metrics.\n",
    "    def plot_isoforms_summary(self):\n",
    "        \"\"\"\n",
    "        Generates a combined plot summarizing:\n",
    "        1. Percentage of genes with multiple isoforms.\n",
    "        2. Frequency distribution of isoforms per gene.\n",
    "        3. Number of genes expressed per cell type.\n",
    "\n",
    "        Returns:\n",
    "        - Combined plot saved as an image.\n",
    "        \"\"\"\n",
    "        ax1 = pw.Brick(figsize=(4, 4))\n",
    "        self._plot_switch_gen_bar(ax1)\n",
    "        ax1.set_title(\"Multiple isoforms genes %\")\n",
    "        ax2 = pw.Brick(figsize=(4, 4))\n",
    "        self._plot_isoforms_frequencies(ax2)\n",
    "        ax2.set_title(\"Frequency of isoforms per gene\")\n",
    "        ax3 = pw.Brick(figsize=(3, 2))\n",
    "        self._plot_genes_cell_type(ax3)\n",
    "        ax3.set_title(\"Nb of genes per cell type\")\n",
    "        return (ax1 | ax2 | ax3).savefig()\n",
    "\n",
    "    # Private method to create a boxplot of genes expressed per cell type.\n",
    "    def _plot_genes_cell_type(self, _ax):\n",
    "        \"\"\"\n",
    "        Plots the number of genes expressed per cell type as a boxplot.\n",
    "\n",
    "        Parameters:\n",
    "        - _ax (Axes): Matplotlib axes to render the plot.\n",
    "        \"\"\"\n",
    "        if _ax is None:\n",
    "            df = pd.DataFrame(np.transpose(self.X), columns=self.obs['cell_type'])\n",
    "            df = df.sum(axis=0).to_frame().reset_index()\n",
    "            df.columns = ['cell_type', 'n_of_genes']\n",
    "            ax = sns.boxplot(x='cell_type', y='n_of_genes', data=df)\n",
    "            ax1 = sns.stripplot(x='cell_type', y='n_of_genes', data=df, color='black', size=3)\n",
    "            ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "            plt.show()\n",
    "        else:\n",
    "            df = pd.DataFrame(np.transpose(self.X), columns=self.obs['cell_type'])\n",
    "            df = df.sum(axis=0).to_frame().reset_index()\n",
    "            df.columns = ['cell_type', 'n_of_genes']\n",
    "            sns.boxplot(x='cell_type', y='n_of_genes', data=df, ax=_ax)\n",
    "            sns.stripplot(x='cell_type', y='n_of_genes', data=df, color='black', size=3, ax=_ax)\n",
    "            _ax.set_xticklabels(_ax.get_xticklabels(), rotation=90)\n",
    "\n",
    "    def plot_genes_cell_type(self):\n",
    "        \"\"\"\n",
    "        Public method to plot the number of genes expressed per cell type.\n",
    "        \"\"\"\n",
    "        self._plot_genes_cell_type(None)\n",
    "\n",
    "    # Method to visualize isoform frequency distribution.\n",
    "    def _plot_isoforms_frequencies(self, _ax):\n",
    "        \"\"\"\n",
    "        Plots the frequency distribution of isoforms per gene.\n",
    "\n",
    "        Parameters:\n",
    "        - _ax (Axes): Matplotlib axes to render the plot.\n",
    "        \"\"\"\n",
    "        if _ax is None:\n",
    "            fig, ax = plt.subplots()\n",
    "            self.gene_counts['transcriptId'].value_counts().plot(\n",
    "                ax=ax, kind='bar', xlabel='number of isoforms per gene', ylabel='quantity of genes'\n",
    "            )\n",
    "        else:\n",
    "            self.gene_counts['transcriptId'].value_counts().plot(\n",
    "                ax=_ax, kind='bar', xlabel='number of isoforms per gene', ylabel='quantity of genes'\n",
    "            )\n",
    "\n",
    "    def plot_isoforms_frequencies(self):\n",
    "        \"\"\"\n",
    "        Public method to plot the frequency distribution of isoforms per gene.\n",
    "        \"\"\"\n",
    "        self._plot_isoforms_frequencies(None)\n",
    "\n",
    "    def plot_switch_gen_bar (self, _ax):\n",
    "        self._plot_switch_gen_bar(None)\n",
    "\n",
    "    def _plot_switch_gen_bar (self, _ax):\n",
    "        if _ax is None:\n",
    "            iso_per_gene = self.gene_counts\n",
    "            x = ['genes']\n",
    "            x1 = ['transcripts']\n",
    "            multiple_iso = sum(iso_per_gene['transcriptId'] > 1)\n",
    "            mono_iso = sum((iso_per_gene['transcriptId'] > 1) == False)\n",
    "            labels = [str(round(1000*multiple_iso/(multiple_iso+mono_iso))/10) + '%', str(round(1000*mono_iso/(multiple_iso+mono_iso))/10) + '%']\n",
    "            fig, ax = plt.subplots()\n",
    "            mult = ax.bar(x, multiple_iso, label=labels[1], color=self.colors[0]) #multiple_iso/(multiple_iso+mono_iso))\n",
    "            mono = ax.bar(x, mono_iso, bottom=multiple_iso, label=labels[0], color=self.colors[0]) #mona_iso/(multiple_iso+mono_iso))\n",
    "            tran = ax.bar(x1, len(self.var['transcriptId']), color=self.colors[2])\n",
    "            ax.text(\n",
    "                ax.patches[0].get_x() + ax.patches[0].get_width() / 2, ax.patches[0].get_height() / 2, labels[0], ha=\"center\", va=\"center\"\n",
    "                )\n",
    "            ax.text(\n",
    "                ax.patches[1].get_x() + ax.patches[1].get_width() / 2, ax.patches[1].get_height() / 2 + ax.patches[0].get_height(), labels[1], ha=\"center\", va=\"center\"\n",
    "                )\n",
    "            plt.legend(['Multiple isoforms', 'Single isoform'])\n",
    "            plt.show()\n",
    "        else:\n",
    "            iso_per_gene = self.gene_counts\n",
    "            x = ['genes']\n",
    "            x1 = ['transcripts']\n",
    "            multiple_iso = sum(iso_per_gene['transcriptId'] > 1)\n",
    "            mono_iso = sum((iso_per_gene['transcriptId'] > 1) == False)\n",
    "            labels = [str(round(1000*multiple_iso/(multiple_iso+mono_iso))/10) + '%', str(round(1000*mono_iso/(multiple_iso+mono_iso))/10) + '%']\n",
    "            mult = _ax.bar(x, multiple_iso, label=labels[1], color=self.colors[0]) #multiple_iso/(multiple_iso+mono_iso))\n",
    "            mono = _ax.bar(x, mono_iso, bottom=multiple_iso, label=labels[0], color=self.colors[1]) #mona_iso/(multiple_iso+mono_iso))\n",
    "            tran = _ax.bar(x1, len(self.var['transcriptId']), color=self.colors[2])\n",
    "            _ax.text(\n",
    "                _ax.patches[0].get_x() + _ax.patches[0].get_width() / 2, _ax.patches[0].get_height() / 2, labels[0], ha=\"center\", va=\"center\"\n",
    "                )\n",
    "            _ax.text(\n",
    "                _ax.patches[1].get_x() + _ax.patches[1].get_width() / 2, _ax.patches[1].get_height() / 2 + _ax.patches[0].get_height(), labels[1], ha=\"center\", va=\"center\"\n",
    "                )\n",
    "            _ax.legend(['Multiple isoforms', 'Single isoform'])\n",
    "            plt.show()\n",
    "\n",
    "    def _trsct_counts_cell_type (self, gene_name, trs_to_show, _ax):\n",
    "        if _ax is None:\n",
    "        # create df with filtered isoform counts and labeled cell types:\n",
    "            df = self._filtered_anndata.to_df().set_index(self._filtered_anndata.obs['cell_type'])\n",
    "            df = df.transpose()\n",
    "            df[['transcriptId', 'geneId']] = self._filtered_anndata.var[['transcriptId', 'geneId']]\n",
    "            gene_iso_count = df[df['geneId'] == gene_name]\n",
    "            gene_iso_count = gene_iso_count.drop('geneId', axis=1).set_index('transcriptId').transpose()\n",
    "            gene_iso_count_long = gene_iso_count.reset_index().melt(id_vars='cell_type', var_name='transcriptId', value_name='count')\n",
    "            if trs_to_show != []:\n",
    "                gene_iso_count_long = gene_iso_count_long[gene_iso_count_long['transcriptId'].isin(trs_to_show)]\n",
    "            g = sns.catplot(x=\"cell_type\", y=\"count\", col=\"transcriptId\", aspect=1, dodge=False, kind=\"violin\", data=gene_iso_count_long, palette=self.colors)\n",
    "            # Set custom facet titles\n",
    "            g.set_titles(col_template=\"{col_name}\", size = 8)\n",
    "            # Remove x ticks\n",
    "            g.set_xticklabels(rotation=90)\n",
    "            g.fig.suptitle(gene_name)\n",
    "            plt.show()\n",
    "        else:\n",
    "            df = self._filtered_anndata.to_df ().set_index(self._filtered_anndata.obs['cell_type'])\n",
    "            df = df.transpose()\n",
    "            df[['transcriptId', 'geneId']] = self._filtered_anndata.var[['transcriptId', 'geneId']]\n",
    "            gene_iso_count = df[df['geneId']== gene_name]\n",
    "            gene_iso_count = gene_iso_count.drop('geneId', axis=1).set_index('transcriptId').transpose()\n",
    "            gene_iso_count_long = gene_iso_count.reset_index().melt(id_vars='cell_type', var_name='transcriptId', value_name='count')\n",
    "            if trs_to_show != []:\n",
    "                gene_iso_count_long = gene_iso_count_long[gene_iso_count_long['transcriptId'].isin(trs_to_show)]\n",
    "            g = sns.catplot(x=\"cell_type\", y=\"count\", col=\"transcriptId\", aspect=1, dodge=False, kind=\"violin\", data=gene_iso_count_long, palette=self.colors)\n",
    "            # Set custom facet titles\n",
    "            g.set_titles(col_template=\"{col_name}\", size = 8)\n",
    "            # Remove x ticks\n",
    "            g.set_xticklabels(rotation=90, labels=self.obs['cell_type'].unique())\n",
    "            g.fig.suptitle(gene_name)\n",
    "            return g\n",
    "\n",
    "    def trsct_counts_cell_type (self, gene_name, trs_to_show = []):\n",
    "        # create df with filtered isoform counts and labeled cell types:\n",
    "        self._trsct_counts_cell_type(gene_name, trs_to_show, None)\n",
    "\n",
    "    def _plot_transcripts_per_cell_type(self, gene_name, trs_to_show, _ax):\n",
    "        if trs_to_show == []:\n",
    "            transcripts_id = self._get_transcripts_from_gene(gene_name)\n",
    "        else:\n",
    "            transcripts_id = trs_to_show\n",
    "        if _ax is None:\n",
    "            fig, ax = plt.subplots()\n",
    "            grouped = self._filtered_anndata.obsm['Iso_prct']\n",
    "            grouped['cell_type'] = self.obs['cell_type']\n",
    "            res = grouped.groupby('cell_type').mean().transpose()\n",
    "            res = res.assign(transcriptId=self._filtered_anndata.var['transcriptId'].to_list())\n",
    "            res = res.assign(geneId=self._filtered_anndata.var['geneId'].to_list())\n",
    "            res = res[res['geneId'] == gene_name].drop(['geneId'], axis=1)\n",
    "            res = res[res['transcriptId'].isin(transcripts_id)]\n",
    "            plot_data = res.set_index('transcriptId').transpose()\n",
    "            plot_data.plot(kind='barh', ax=ax, stacked=True, color=self.colors).legend(loc='center left',bbox_to_anchor=(1.0, 1.0))\n",
    "            plt.legend(self.get_transcripts_common_names(transcripts_id), loc=\"upper left\", bbox_to_anchor=(1, 1))\n",
    "            #plt.legend(self.get_transcripts_common_names(trs_to_show))\n",
    "            plt.ylabel('Cell type')\n",
    "        else:\n",
    "            grouped = self._filtered_anndata.obsm['Iso_prct']\n",
    "            grouped['cell_type'] = self.obs['cell_type']\n",
    "            res = grouped.groupby('cell_type').mean().transpose()\n",
    "            res = res.assign(transcriptId=self._filtered_anndata.var['transcriptId'].to_list())\n",
    "            res = res.assign(geneId=self._filtered_anndata.var['geneId'].to_list())\n",
    "            res = res[res['geneId'] == gene_name].drop(['geneId'], axis=1)\n",
    "            res = res[res['transcriptId'].isin(transcripts_id)]\n",
    "            plot_data = res.set_index('transcriptId').transpose()\n",
    "            plot_data.plot(kind='barh', stacked=True, ax=_ax, color=self.colors).legend(loc='center left',bbox_to_anchor=(1.0, 1.0))\n",
    "\n",
    "    def plot_transcripts_per_cell_type(self, gene_name, trs_to_show = []):\n",
    "        self._plot_transcripts_per_cell_type(gene_name, trs_to_show, None)\n",
    "\n",
    "    def _get_transcripts_from_gene(self, gene_name):\n",
    "        elems = self._filtered_anndata.var\n",
    "        return elems[elems['geneId'] == gene_name]['transcriptId'].to_list()\n",
    "\n",
    "    def _draw_transcripts_list_from_gene(self, gene_name, trs_to_show, _ax, colors=None):\n",
    "        if trs_to_show == []:\n",
    "            transcripts_id = self._get_transcripts_from_gene(gene_name)\n",
    "        else:\n",
    "            transcripts_id = trs_to_show\n",
    "        exons = []\n",
    "        directions = []\n",
    "        for tr in transcripts_id:\n",
    "            t, d = self._get_coord_from_tscrpt_id(tr)\n",
    "            exons += [t]\n",
    "            directions += [d]\n",
    "        if colors is None:\n",
    "            colors = []\n",
    "            for i in range(len(exons)):\n",
    "                colors.append(self.colors[i % len(self.colors)])\n",
    "        def get_limits(ex, dir):\n",
    "            start = sys.maxsize\n",
    "            end = -sys.maxsize\n",
    "            for (e, d) in zip(ex, dir):\n",
    "                if d == 1:\n",
    "                    start = min(start,e[0][1])\n",
    "                    end = max(end, e[-1][0])\n",
    "                else:\n",
    "                    start = min(start,e[-1][1])\n",
    "                    end = max(end, e[0][0])\n",
    "            return (start, end)\n",
    "        plt.axes()\n",
    "        plt.xlim((-0.1, 1.1))\n",
    "        plt.ylim((0.1 - 0.5 *  len(exons), 0.3))\n",
    "        plt.margins(0.2)\n",
    "        plt.axis('off')\n",
    "        fig = plt.gcf()\n",
    "        fig.set_size_inches(20, len(exons) * 2)\n",
    "        i = 0\n",
    "        (start, end) = get_limits(exons, directions)\n",
    "        for (ex, di, co, name) in zip(exons, directions, colors, transcripts_id):\n",
    "            self._draw_transcript(ex, di, co, name, offset= -0.5 * i, start_override=start, end_override=end, no_render=True)\n",
    "            i+=1\n",
    "        if _ax is None:\n",
    "            plt.show()\n",
    "        else:\n",
    "            return plt\n",
    "\n",
    "    def draw_transcripts_list_from_gene(self, gene_name, colors=None):\n",
    "        self._draw_transcripts_list_from_gene(gene_name, [], None, colors)\n",
    "\n",
    "    def draw_gene_summary(self, gene_name, trs_to_show = []):\n",
    "        ax1 = pw.Brick(figsize=(12,4))\n",
    "        pw.overwrite_axisgrid()\n",
    "        fg = self._trsct_counts_cell_type(gene_name, trs_to_show, ax1)\n",
    "        ax4 = pw.load_seaborngrid(fg)\n",
    "        ax1.set_title(\"Transcripts count per cell type\")\n",
    "        ax2 = pw.Brick(figsize=(12,4))\n",
    "        self._plot_transcripts_per_cell_type(gene_name, trs_to_show, ax2)\n",
    "        ax2.set_title(\"Transcripts count per cell type\")\n",
    "        ax3 = pw.Brick(figsize=(12,4))\n",
    "        self._draw_transcripts_list_from_gene(gene_name, trs_to_show, ax3)\n",
    "        ax3.set_title(\"Transcripts list\")\n",
    "        return (ax4/(ax2/ax3)).savefig()\n",
    "\n",
    "    # More detailed functions omitted for brevity."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
